{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "015b5449",
   "metadata": {
    "papermill": {
     "duration": 0.00634,
     "end_time": "2025-01-06T21:09:42.058571",
     "exception": false,
     "start_time": "2025-01-06T21:09:42.052231",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Ensembl genes table extraction EDA\n",
    "\n",
    "This notebook is useful for development as well as exploratory data analysis on the extracted tables.\n",
    "It is currently automically executed and saved as part of exports using `papermill`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3345ae4a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:09:42.071582Z",
     "iopub.status.busy": "2025-01-06T21:09:42.071111Z",
     "iopub.status.idle": "2025-01-06T21:09:42.089006Z",
     "shell.execute_reply": "2025-01-06T21:09:42.088326Z"
    },
    "papermill": {
     "duration": 0.026082,
     "end_time": "2025-01-06T21:09:42.090470",
     "exception": false,
     "start_time": "2025-01-06T21:09:42.064388",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08df3755",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:09:42.103589Z",
     "iopub.status.busy": "2025-01-06T21:09:42.103140Z",
     "iopub.status.idle": "2025-01-06T21:09:44.061638Z",
     "shell.execute_reply": "2025-01-06T21:09:44.060913Z"
    },
    "papermill": {
     "duration": 1.966633,
     "end_time": "2025-01-06T21:09:44.063248",
     "exception": false,
     "start_time": "2025-01-06T21:09:42.096615",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from ensembl_genes import ensembl_genes\n",
    "from bioregistry import normalize_prefix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a140647f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:09:44.076770Z",
     "iopub.status.busy": "2025-01-06T21:09:44.076224Z",
     "iopub.status.idle": "2025-01-06T21:09:44.089358Z",
     "shell.execute_reply": "2025-01-06T21:09:44.088813Z"
    },
    "papermill": {
     "duration": 0.021217,
     "end_time": "2025-01-06T21:09:44.090724",
     "exception": false,
     "start_time": "2025-01-06T21:09:44.069507",
     "status": "completed"
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# parameters cell\n",
    "species = \"human\"\n",
    "release = \"111\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c0d7a192",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:09:44.103980Z",
     "iopub.status.busy": "2025-01-06T21:09:44.103456Z",
     "iopub.status.idle": "2025-01-06T21:09:44.116939Z",
     "shell.execute_reply": "2025-01-06T21:09:44.116246Z"
    },
    "papermill": {
     "duration": 0.021611,
     "end_time": "2025-01-06T21:09:44.118460",
     "exception": false,
     "start_time": "2025-01-06T21:09:44.096849",
     "status": "completed"
    },
    "tags": [
     "injected-parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# Parameters\n",
    "species = \"mouse\"\n",
    "release = \"113\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5e6b8db2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:09:44.132150Z",
     "iopub.status.busy": "2025-01-06T21:09:44.131663Z",
     "iopub.status.idle": "2025-01-06T21:09:44.221194Z",
     "shell.execute_reply": "2025-01-06T21:09:44.220587Z"
    },
    "papermill": {
     "duration": 0.097848,
     "end_time": "2025-01-06T21:09:44.222536",
     "exception": false,
     "start_time": "2025-01-06T21:09:44.124688",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mysql+mysqlconnector://anonymous@ensembldb.ensembl.org:3306/mus_musculus_core_113_39'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg = ensembl_genes.Ensembl_Gene_Queries(release=release, species=species)\n",
    "ensg.connection_url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0dea0121",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:09:44.236766Z",
     "iopub.status.busy": "2025-01-06T21:09:44.236162Z",
     "iopub.status.idle": "2025-01-06T21:09:44.251363Z",
     "shell.execute_reply": "2025-01-06T21:09:44.250700Z"
    },
    "papermill": {
     "duration": 0.023757,
     "end_time": "2025-01-06T21:09:44.252774",
     "exception": false,
     "start_time": "2025-01-06T21:09:44.229017",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mus_musculus_core_113_39'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "database = ensg.database\n",
    "database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2197def",
   "metadata": {
    "papermill": {
     "duration": 0.006121,
     "end_time": "2025-01-06T21:09:44.265430",
     "exception": false,
     "start_time": "2025-01-06T21:09:44.259309",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## Extract data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c90c658",
   "metadata": {
    "papermill": {
     "duration": 0.00611,
     "end_time": "2025-01-06T21:09:44.277619",
     "exception": false,
     "start_time": "2025-01-06T21:09:44.271509",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## gene attrib counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ce76b040",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:09:44.291306Z",
     "iopub.status.busy": "2025-01-06T21:09:44.290805Z",
     "iopub.status.idle": "2025-01-06T21:09:47.333389Z",
     "shell.execute_reply": "2025-01-06T21:09:47.332729Z"
    },
    "papermill": {
     "duration": 3.051056,
     "end_time": "2025-01-06T21:09:47.334720",
     "exception": false,
     "start_time": "2025-01-06T21:09:44.283664",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>attrib_type_id</th>\n",
       "      <th>code</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>attrib_type_count</th>\n",
       "      <th>attrib_type_examples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>142</td>\n",
       "      <td>GeneGC</td>\n",
       "      <td>Gene GC</td>\n",
       "      <td>Percentage GC content for this gene</td>\n",
       "      <td>78298</td>\n",
       "      <td>52.38, 41.29, 45.11, 46.08, 58.82, 45.96, 38.3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>name</td>\n",
       "      <td>Name</td>\n",
       "      <td>Alternative/long name</td>\n",
       "      <td>72073</td>\n",
       "      <td>RP24-408N5.1, Clec12b, RP23-454G7.1, Slc35e2, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>395</td>\n",
       "      <td>xref_id</td>\n",
       "      <td>Xref ID</td>\n",
       "      <td>ID of associated database reference</td>\n",
       "      <td>48559</td>\n",
       "      <td>OTTMUSG00000050759, OTTMUSG00000056465, OTTMUS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>538</td>\n",
       "      <td>legacy_biotype</td>\n",
       "      <td>Legacy biotype</td>\n",
       "      <td>Obsolete biotype previously assigned to this E...</td>\n",
       "      <td>32957</td>\n",
       "      <td>antisense, lincRNA, sense_overlapping, process...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>54</td>\n",
       "      <td>remark</td>\n",
       "      <td>Remark</td>\n",
       "      <td>Annotation remark</td>\n",
       "      <td>23152</td>\n",
       "      <td>TAGENE_gene, D1Ertd622e, (Dodds et al., J Biol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>380</td>\n",
       "      <td>havana_cv</td>\n",
       "      <td>Havana CV term</td>\n",
       "      <td>Controlled vocabulary terms from Havana</td>\n",
       "      <td>11532</td>\n",
       "      <td>overlapping locus, overlaps pseudogene, confir...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>382</td>\n",
       "      <td>NoTransRefError</td>\n",
       "      <td>No translations due to reference error</td>\n",
       "      <td>This gene is believed to include protein codin...</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>embl_acc</td>\n",
       "      <td>European Nucleotide Archive (was EMBL) accession</td>\n",
       "      <td>ENA</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>status</td>\n",
       "      <td>Status</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>synonym</td>\n",
       "      <td>Synonym</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5</td>\n",
       "      <td>type</td>\n",
       "      <td>Type of feature</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>6</td>\n",
       "      <td>toplevel</td>\n",
       "      <td>Top Level</td>\n",
       "      <td>Top Level Non-Redundant Sequence Region</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>7</td>\n",
       "      <td>GeneCount</td>\n",
       "      <td>Gene Count</td>\n",
       "      <td>Total Number of Genes</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10</td>\n",
       "      <td>SNPCount</td>\n",
       "      <td>Short Variants</td>\n",
       "      <td>Total Number of SNPs</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>11</td>\n",
       "      <td>codon_table</td>\n",
       "      <td>Codon Table</td>\n",
       "      <td>Alternate codon table</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    attrib_type_id             code  \\\n",
       "0              142           GeneGC   \n",
       "1                4             name   \n",
       "2              395          xref_id   \n",
       "3              538   legacy_biotype   \n",
       "4               54           remark   \n",
       "5              380        havana_cv   \n",
       "6              382  NoTransRefError   \n",
       "7                1         embl_acc   \n",
       "8                2           status   \n",
       "9                3          synonym   \n",
       "10               5             type   \n",
       "11               6         toplevel   \n",
       "12               7        GeneCount   \n",
       "13              10         SNPCount   \n",
       "14              11      codon_table   \n",
       "\n",
       "                                                name  \\\n",
       "0                                            Gene GC   \n",
       "1                                               Name   \n",
       "2                                            Xref ID   \n",
       "3                                     Legacy biotype   \n",
       "4                                             Remark   \n",
       "5                                     Havana CV term   \n",
       "6             No translations due to reference error   \n",
       "7   European Nucleotide Archive (was EMBL) accession   \n",
       "8                                             Status   \n",
       "9                                            Synonym   \n",
       "10                                   Type of feature   \n",
       "11                                         Top Level   \n",
       "12                                        Gene Count   \n",
       "13                                    Short Variants   \n",
       "14                                       Codon Table   \n",
       "\n",
       "                                          description  attrib_type_count  \\\n",
       "0                 Percentage GC content for this gene              78298   \n",
       "1                               Alternative/long name              72073   \n",
       "2                 ID of associated database reference              48559   \n",
       "3   Obsolete biotype previously assigned to this E...              32957   \n",
       "4                                   Annotation remark              23152   \n",
       "5             Controlled vocabulary terms from Havana              11532   \n",
       "6   This gene is believed to include protein codin...                  5   \n",
       "7                                                 ENA                  0   \n",
       "8                                                None                  0   \n",
       "9                                                None                  0   \n",
       "10                                               None                  0   \n",
       "11            Top Level Non-Redundant Sequence Region                  0   \n",
       "12                              Total Number of Genes                  0   \n",
       "13                               Total Number of SNPs                  0   \n",
       "14                              Alternate codon table                  0   \n",
       "\n",
       "                                 attrib_type_examples  \n",
       "0   52.38, 41.29, 45.11, 46.08, 58.82, 45.96, 38.3...  \n",
       "1   RP24-408N5.1, Clec12b, RP23-454G7.1, Slc35e2, ...  \n",
       "2   OTTMUSG00000050759, OTTMUSG00000056465, OTTMUS...  \n",
       "3   antisense, lincRNA, sense_overlapping, process...  \n",
       "4   TAGENE_gene, D1Ertd622e, (Dodds et al., J Biol...  \n",
       "5   overlapping locus, overlaps pseudogene, confir...  \n",
       "6                                                   1  \n",
       "7                                                None  \n",
       "8                                                None  \n",
       "9                                                None  \n",
       "10                                               None  \n",
       "11                                               None  \n",
       "12                                               None  \n",
       "13                                               None  \n",
       "14                                               None  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.run_query(\"gene_attrib_counts\").head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bc589fc",
   "metadata": {
    "papermill": {
     "duration": 0.00618,
     "end_time": "2025-01-06T21:09:47.347650",
     "exception": false,
     "start_time": "2025-01-06T21:09:47.341470",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "366632f2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:09:47.361364Z",
     "iopub.status.busy": "2025-01-06T21:09:47.360969Z",
     "iopub.status.idle": "2025-01-06T21:13:03.175373Z",
     "shell.execute_reply": "2025-01-06T21:13:03.174637Z"
    },
    "papermill": {
     "duration": 195.829779,
     "end_time": "2025-01-06T21:13:03.183639",
     "exception": false,
     "start_time": "2025-01-06T21:09:47.353860",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_version</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_symbol_source_db</th>\n",
       "      <th>gene_symbol_source_id</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>ensembl_source</th>\n",
       "      <th>ensembl_created_date</th>\n",
       "      <th>ensembl_modified_date</th>\n",
       "      <th>coord_system_version</th>\n",
       "      <th>...</th>\n",
       "      <th>seq_region_start</th>\n",
       "      <th>seq_region_end</th>\n",
       "      <th>seq_region_strand</th>\n",
       "      <th>primary_assembly</th>\n",
       "      <th>lrg_gene_id</th>\n",
       "      <th>mhc</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>gene_description_source_db</th>\n",
       "      <th>gene_description_source_id</th>\n",
       "      <th>ensembl_representative_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>5</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:95773</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>2004-11-18 15:05:57</td>\n",
       "      <td>2020-08-10 14:30:16</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>108014596</td>\n",
       "      <td>108053462</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>G protein subunit alpha i3</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>MGI:95773</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "      <td>16</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:1860484</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>2006-06-28 13:29:12</td>\n",
       "      <td>2020-08-10 14:24:33</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>76881507</td>\n",
       "      <td>76897229</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>probasin</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>MGI:1860484</td>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000028</td>\n",
       "      <td>16</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:1338073</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>2007-06-26 09:54:29</td>\n",
       "      <td>2020-08-10 14:17:14</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>18599197</td>\n",
       "      <td>18630737</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>cell division cycle 45</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>MGI:1338073</td>\n",
       "      <td>ENSMUSG00000000028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000031</td>\n",
       "      <td>20</td>\n",
       "      <td>H19</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:95891</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>2006-02-02 11:27:21</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>142129262</td>\n",
       "      <td>142133957</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>H19, imprinted maternally expressed transcript</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>MGI:95891</td>\n",
       "      <td>ENSMUSG00000000031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>18</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:1340042</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>2006-08-18 16:16:32</td>\n",
       "      <td>2020-08-10 14:22:47</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>159865521</td>\n",
       "      <td>160041209</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>Scm polycomb group protein like 2</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>MGI:1340042</td>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ensembl_gene_id  ensembl_gene_version gene_symbol gene_symbol_source_db  \\\n",
       "0  ENSMUSG00000000001                     5       Gnai3                   MGI   \n",
       "1  ENSMUSG00000000003                    16        Pbsn                   MGI   \n",
       "2  ENSMUSG00000000028                    16       Cdc45                   MGI   \n",
       "3  ENSMUSG00000000031                    20         H19                   MGI   \n",
       "4  ENSMUSG00000000037                    18       Scml2                   MGI   \n",
       "\n",
       "  gene_symbol_source_id    gene_biotype  ensembl_source ensembl_created_date  \\\n",
       "0             MGI:95773  protein_coding  ensembl_havana  2004-11-18 15:05:57   \n",
       "1           MGI:1860484  protein_coding  ensembl_havana  2006-06-28 13:29:12   \n",
       "2           MGI:1338073  protein_coding  ensembl_havana  2007-06-26 09:54:29   \n",
       "3             MGI:95891          lncRNA  ensembl_havana  2006-02-02 11:27:21   \n",
       "4           MGI:1340042  protein_coding  ensembl_havana  2006-08-18 16:16:32   \n",
       "\n",
       "  ensembl_modified_date coord_system_version  ... seq_region_start  \\\n",
       "0   2020-08-10 14:30:16               GRCm39  ...        108014596   \n",
       "1   2020-08-10 14:24:33               GRCm39  ...         76881507   \n",
       "2   2020-08-10 14:17:14               GRCm39  ...         18599197   \n",
       "3   2024-05-06 19:10:14               GRCm39  ...        142129262   \n",
       "4   2020-08-10 14:22:47               GRCm39  ...        159865521   \n",
       "\n",
       "  seq_region_end seq_region_strand  primary_assembly  lrg_gene_id   mhc  \\\n",
       "0      108053462                -1              True          NaN  None   \n",
       "1       76897229                -1              True          NaN  None   \n",
       "2       18630737                -1              True          NaN  None   \n",
       "3      142133957                -1              True          NaN  None   \n",
       "4      160041209                 1              True          NaN  None   \n",
       "\n",
       "                                 gene_description gene_description_source_db  \\\n",
       "0                      G protein subunit alpha i3                 MGI Symbol   \n",
       "1                                        probasin                 MGI Symbol   \n",
       "2                          cell division cycle 45                 MGI Symbol   \n",
       "3  H19, imprinted maternally expressed transcript                 MGI Symbol   \n",
       "4               Scm polycomb group protein like 2                 MGI Symbol   \n",
       "\n",
       "  gene_description_source_id ensembl_representative_gene_id  \n",
       "0                  MGI:95773             ENSMUSG00000000001  \n",
       "1                MGI:1860484             ENSMUSG00000000003  \n",
       "2                MGI:1338073             ENSMUSG00000000028  \n",
       "3                  MGI:95891             ENSMUSG00000000031  \n",
       "4                MGI:1340042             ENSMUSG00000000037  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f15c5ebd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.198492Z",
     "iopub.status.busy": "2025-01-06T21:13:03.197936Z",
     "iopub.status.idle": "2025-01-06T21:13:03.234497Z",
     "shell.execute_reply": "2025-01-06T21:13:03.233803Z"
    },
    "papermill": {
     "duration": 0.045407,
     "end_time": "2025-01-06T21:13:03.235848",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.190441",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_version</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_symbol_source_db</th>\n",
       "      <th>gene_symbol_source_id</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>ensembl_source</th>\n",
       "      <th>ensembl_created_date</th>\n",
       "      <th>ensembl_modified_date</th>\n",
       "      <th>coord_system_version</th>\n",
       "      <th>...</th>\n",
       "      <th>seq_region_start</th>\n",
       "      <th>seq_region_end</th>\n",
       "      <th>seq_region_strand</th>\n",
       "      <th>primary_assembly</th>\n",
       "      <th>lrg_gene_id</th>\n",
       "      <th>mhc</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>gene_description_source_db</th>\n",
       "      <th>gene_description_source_id</th>\n",
       "      <th>ensembl_representative_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13860</th>\n",
       "      <td>ENSMUSG00000047061</td>\n",
       "      <td>7</td>\n",
       "      <td>ENSMUSG00000047061</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TEC</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>2018-12-03 13:41:15</td>\n",
       "      <td>2020-08-10 14:05:17</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>45232168</td>\n",
       "      <td>45234698</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>TEC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000047061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20035</th>\n",
       "      <td>ENSMUSG00000072963</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000072963</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>2018-11-30 11:32:45</td>\n",
       "      <td>2020-08-10 14:22:30</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>53345631</td>\n",
       "      <td>53347937</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000072963</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ensembl_gene_id  ensembl_gene_version         gene_symbol  \\\n",
       "13860  ENSMUSG00000047061                     7  ENSMUSG00000047061   \n",
       "20035  ENSMUSG00000072963                     3  ENSMUSG00000072963   \n",
       "\n",
       "      gene_symbol_source_db gene_symbol_source_id gene_biotype  \\\n",
       "13860                  None                  None          TEC   \n",
       "20035                  None                  None       lncRNA   \n",
       "\n",
       "       ensembl_source ensembl_created_date ensembl_modified_date  \\\n",
       "13860  ensembl_havana  2018-12-03 13:41:15   2020-08-10 14:05:17   \n",
       "20035  ensembl_havana  2018-11-30 11:32:45   2020-08-10 14:22:30   \n",
       "\n",
       "      coord_system_version  ... seq_region_start seq_region_end  \\\n",
       "13860               GRCm39  ...         45232168       45234698   \n",
       "20035               GRCm39  ...         53345631       53347937   \n",
       "\n",
       "      seq_region_strand  primary_assembly  lrg_gene_id   mhc  \\\n",
       "13860                 1              True          NaN  None   \n",
       "20035                -1              True          NaN  None   \n",
       "\n",
       "       gene_description gene_description_source_db gene_description_source_id  \\\n",
       "13860               TEC                        NaN                        NaN   \n",
       "20035  novel transcript                        NaN                        NaN   \n",
       "\n",
       "      ensembl_representative_gene_id  \n",
       "13860             ENSMUSG00000047061  \n",
       "20035             ENSMUSG00000072963  \n",
       "\n",
       "[2 rows x 23 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clone-based genes no longer get a symbol and are filled with the stable ID\n",
    "# https://www.ensembl.info/2021/03/15/retirement-of-clone-based-gene-names/\n",
    "ensg.gene_df.query(\"gene_symbol == ensembl_gene_id\").head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "09e2788a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.251098Z",
     "iopub.status.busy": "2025-01-06T21:13:03.250590Z",
     "iopub.status.idle": "2025-01-06T21:13:03.351674Z",
     "shell.execute_reply": "2025-01-06T21:13:03.350948Z"
    },
    "papermill": {
     "duration": 0.10992,
     "end_time": "2025-01-06T21:13:03.353019",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.243099",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_symbol_source_db</th>\n",
       "      <th>EntrezGene</th>\n",
       "      <th>HGNC</th>\n",
       "      <th>MGI</th>\n",
       "      <th>missing (clone-based)</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_source</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RefSeq</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4924</td>\n",
       "      <td>51</td>\n",
       "      <td>4978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl_havana</th>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>22435</td>\n",
       "      <td>38</td>\n",
       "      <td>22508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>havana</th>\n",
       "      <td>87</td>\n",
       "      <td>3</td>\n",
       "      <td>26270</td>\n",
       "      <td>454</td>\n",
       "      <td>26814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>havana_tagene</th>\n",
       "      <td>165</td>\n",
       "      <td>0</td>\n",
       "      <td>1269</td>\n",
       "      <td>21340</td>\n",
       "      <td>22774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mirbase</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1187</td>\n",
       "      <td>0</td>\n",
       "      <td>1187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>290</td>\n",
       "      <td>3</td>\n",
       "      <td>56122</td>\n",
       "      <td>21883</td>\n",
       "      <td>78298</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_symbol_source_db  EntrezGene  HGNC    MGI  missing (clone-based)    All\n",
       "ensembl_source                                                              \n",
       "RefSeq                          0     0     37                      0     37\n",
       "ensembl                         3     0   4924                     51   4978\n",
       "ensembl_havana                 35     0  22435                     38  22508\n",
       "havana                         87     3  26270                    454  26814\n",
       "havana_tagene                 165     0   1269                  21340  22774\n",
       "mirbase                         0     0   1187                      0   1187\n",
       "All                           290     3  56122                  21883  78298"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# which external database the gene symbol derives from versus the ensembl source\n",
    "pd.crosstab(\n",
    "    ensg.gene_df.ensembl_source,\n",
    "    ensg.gene_df.gene_symbol_source_db.fillna(\"missing (clone-based)\"),\n",
    "    margins=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "90ff6f84",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.368739Z",
     "iopub.status.busy": "2025-01-06T21:13:03.368107Z",
     "iopub.status.idle": "2025-01-06T21:13:03.392607Z",
     "shell.execute_reply": "2025-01-06T21:13:03.392000Z"
    },
    "papermill": {
     "duration": 0.033422,
     "end_time": "2025-01-06T21:13:03.393943",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.360521",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chromosome    78239\n",
       "scaffold         59\n",
       "Name: coord_system, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.gene_df.coord_system.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "72df921b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.408890Z",
     "iopub.status.busy": "2025-01-06T21:13:03.408694Z",
     "iopub.status.idle": "2025-01-06T21:13:03.433893Z",
     "shell.execute_reply": "2025-01-06T21:13:03.432767Z"
    },
    "papermill": {
     "duration": 0.034255,
     "end_time": "2025-01-06T21:13:03.435332",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.401077",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "lncRNA                                32957\n",
       "protein_coding                        21748\n",
       "processed_pseudogene                   9329\n",
       "TEC                                    3220\n",
       "miRNA                                  2206\n",
       "unprocessed_pseudogene                 2158\n",
       "snoRNA                                 1507\n",
       "snRNA                                  1381\n",
       "transcribed_unprocessed_pseudogene      995\n",
       "transcribed_processed_pseudogene        949\n",
       "Name: gene_biotype, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.gene_df.gene_biotype.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6381c484",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.452000Z",
     "iopub.status.busy": "2025-01-06T21:13:03.451507Z",
     "iopub.status.idle": "2025-01-06T21:13:03.516544Z",
     "shell.execute_reply": "2025-01-06T21:13:03.515787Z"
    },
    "papermill": {
     "duration": 0.07472,
     "end_time": "2025-01-06T21:13:03.517864",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.443144",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>primary_assembly</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coord_system</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chromosome</th>\n",
       "      <td>0</td>\n",
       "      <td>78239</td>\n",
       "      <td>78239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scaffold</th>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>59</td>\n",
       "      <td>78239</td>\n",
       "      <td>78298</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "primary_assembly  False   True    All\n",
       "coord_system                         \n",
       "chromosome            0  78239  78239\n",
       "scaffold             59      0     59\n",
       "All                  59  78239  78298"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(ensg.gene_df.coord_system, ensg.gene_df.primary_assembly, margins=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a154bcfb",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.534214Z",
     "iopub.status.busy": "2025-01-06T21:13:03.533812Z",
     "iopub.status.idle": "2025-01-06T21:13:03.549843Z",
     "shell.execute_reply": "2025-01-06T21:13:03.549325Z"
    },
    "papermill": {
     "duration": 0.025366,
     "end_time": "2025-01-06T21:13:03.551124",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.525758",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: mhc, dtype: int64)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.gene_df.mhc.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7b9a7058",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.567285Z",
     "iopub.status.busy": "2025-01-06T21:13:03.566840Z",
     "iopub.status.idle": "2025-01-06T21:13:03.582361Z",
     "shell.execute_reply": "2025-01-06T21:13:03.581717Z"
    },
    "papermill": {
     "duration": 0.025139,
     "end_time": "2025-01-06T21:13:03.583726",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.558587",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "78298"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ensg.gene_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d76eff",
   "metadata": {
    "papermill": {
     "duration": 0.007285,
     "end_time": "2025-01-06T21:13:03.598903",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.591618",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## alternative gene alleles\n",
    "\n",
    "Related:\n",
    "\n",
    "- [OTP: Origin of genes_with_non_reference_ensembl_ids.tsv](https://github.com/opentargets/platform/issues/702)\n",
    "- [biostars: map between different assemblies of one ensembl release](https://www.biostars.org/p/143956/)\n",
    "- using `attrib_type.code = \"non_ref\"` for `primary_assembly` doesn't appear to return any results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "449b4769",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.614858Z",
     "iopub.status.busy": "2025-01-06T21:13:03.614377Z",
     "iopub.status.idle": "2025-01-06T21:13:03.634907Z",
     "shell.execute_reply": "2025-01-06T21:13:03.634257Z"
    },
    "papermill": {
     "duration": 0.030024,
     "end_time": "2025-01-06T21:13:03.636215",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.606191",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs_allele_group</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>ensembl_created_date</th>\n",
       "      <th>seq_region</th>\n",
       "      <th>primary_assembly</th>\n",
       "      <th>alt_allele_group_id</th>\n",
       "      <th>alt_allele_attrib</th>\n",
       "      <th>alt_allele_is_representative</th>\n",
       "      <th>ensembl_representative_gene_id</th>\n",
       "      <th>is_representative_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0610005C13Rik</td>\n",
       "      <td>ENSMUSG00000109644</td>\n",
       "      <td>0610005C13Rik</td>\n",
       "      <td>2007-12-11 14:58:31</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSMUSG00000109644</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0610006L08Rik</td>\n",
       "      <td>ENSMUSG00000108652</td>\n",
       "      <td>0610006L08Rik</td>\n",
       "      <td>2015-07-06 11:47:18</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSMUSG00000108652</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0610009E02Rik</td>\n",
       "      <td>ENSMUSG00000086714</td>\n",
       "      <td>0610009E02Rik</td>\n",
       "      <td>2005-06-29 11:35:32</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSMUSG00000086714</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0610009L18Rik</td>\n",
       "      <td>ENSMUSG00000043644</td>\n",
       "      <td>0610009L18Rik</td>\n",
       "      <td>2004-08-20 10:27:42</td>\n",
       "      <td>11</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSMUSG00000043644</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0610010K14Rik</td>\n",
       "      <td>ENSMUSG00000020831</td>\n",
       "      <td>0610010K14Rik</td>\n",
       "      <td>2003-12-09 12:39:50</td>\n",
       "      <td>11</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSMUSG00000020831</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  rs_allele_group     ensembl_gene_id    gene_symbol ensembl_created_date  \\\n",
       "0   0610005C13Rik  ENSMUSG00000109644  0610005C13Rik  2007-12-11 14:58:31   \n",
       "1   0610006L08Rik  ENSMUSG00000108652  0610006L08Rik  2015-07-06 11:47:18   \n",
       "2   0610009E02Rik  ENSMUSG00000086714  0610009E02Rik  2005-06-29 11:35:32   \n",
       "3   0610009L18Rik  ENSMUSG00000043644  0610009L18Rik  2004-08-20 10:27:42   \n",
       "4   0610010K14Rik  ENSMUSG00000020831  0610010K14Rik  2003-12-09 12:39:50   \n",
       "\n",
       "  seq_region  primary_assembly alt_allele_group_id alt_allele_attrib  \\\n",
       "0          7              True                None              None   \n",
       "1          7              True                None              None   \n",
       "2          2              True                None              None   \n",
       "3         11              True                None              None   \n",
       "4         11              True                None              None   \n",
       "\n",
       "   alt_allele_is_representative ensembl_representative_gene_id  \\\n",
       "0                         False             ENSMUSG00000109644   \n",
       "1                         False             ENSMUSG00000108652   \n",
       "2                         False             ENSMUSG00000086714   \n",
       "3                         False             ENSMUSG00000043644   \n",
       "4                         False             ENSMUSG00000020831   \n",
       "\n",
       "   is_representative_gene  \n",
       "0                    True  \n",
       "1                    True  \n",
       "2                    True  \n",
       "3                    True  \n",
       "4                    True  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.representative_gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5a96079c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:03.652529Z",
     "iopub.status.busy": "2025-01-06T21:13:03.652316Z",
     "iopub.status.idle": "2025-01-06T21:13:06.339509Z",
     "shell.execute_reply": "2025-01-06T21:13:06.338844Z"
    },
    "papermill": {
     "duration": 2.696874,
     "end_time": "2025-01-06T21:13:06.340901",
     "exception": false,
     "start_time": "2025-01-06T21:13:03.644027",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>attrib_type_id</th>\n",
       "      <th>value</th>\n",
       "      <th>attrib_type_id</th>\n",
       "      <th>code</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id, attrib_type_id, value, attrib_type_id, code, name, description]\n",
       "Index: []"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# looks like non_ref isn't set for human genes\n",
    "query = '''\n",
    "SELECT *\n",
    "FROM gene_attrib\n",
    "LEFT JOIN attrib_type\n",
    "  ON gene_attrib.attrib_type_id = attrib_type.attrib_type_id\n",
    "WHERE attrib_type.code = \"non_ref\"\n",
    "LIMIT 5\n",
    "'''\n",
    "pd.read_sql(sql=query, con=ensg.connection_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8f5122c1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:06.359019Z",
     "iopub.status.busy": "2025-01-06T21:13:06.358556Z",
     "iopub.status.idle": "2025-01-06T21:13:06.374703Z",
     "shell.execute_reply": "2025-01-06T21:13:06.374053Z"
    },
    "papermill": {
     "duration": 0.02621,
     "end_time": "2025-01-06T21:13:06.375945",
     "exception": false,
     "start_time": "2025-01-06T21:13:06.349735",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: alt_allele_attrib, dtype: int64)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.representative_gene_df.alt_allele_attrib.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "352ec809",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:06.393481Z",
     "iopub.status.busy": "2025-01-06T21:13:06.392931Z",
     "iopub.status.idle": "2025-01-06T21:13:06.417437Z",
     "shell.execute_reply": "2025-01-06T21:13:06.416780Z"
    },
    "papermill": {
     "duration": 0.034481,
     "end_time": "2025-01-06T21:13:06.418742",
     "exception": false,
     "start_time": "2025-01-06T21:13:06.384261",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs_allele_group</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>ensembl_created_date</th>\n",
       "      <th>seq_region</th>\n",
       "      <th>primary_assembly</th>\n",
       "      <th>alt_allele_group_id</th>\n",
       "      <th>alt_allele_attrib</th>\n",
       "      <th>alt_allele_is_representative</th>\n",
       "      <th>ensembl_representative_gene_id</th>\n",
       "      <th>is_representative_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>1600029O15Rik</td>\n",
       "      <td>ENSMUSG00000121691</td>\n",
       "      <td>1600029O15Rik</td>\n",
       "      <td>2023-02-25 19:58:01</td>\n",
       "      <td>9</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSMUSG00000057818</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>1700016A09Rik</td>\n",
       "      <td>ENSMUSG00000123653</td>\n",
       "      <td>1700016A09Rik</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSMUSG00000102737</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    rs_allele_group     ensembl_gene_id    gene_symbol ensembl_created_date  \\\n",
       "76    1600029O15Rik  ENSMUSG00000121691  1600029O15Rik  2023-02-25 19:58:01   \n",
       "170   1700016A09Rik  ENSMUSG00000123653  1700016A09Rik  2024-05-06 19:10:14   \n",
       "\n",
       "    seq_region  primary_assembly alt_allele_group_id alt_allele_attrib  \\\n",
       "76           9              True                None              None   \n",
       "170          8              True                None              None   \n",
       "\n",
       "     alt_allele_is_representative ensembl_representative_gene_id  \\\n",
       "76                          False             ENSMUSG00000057818   \n",
       "170                         False             ENSMUSG00000102737   \n",
       "\n",
       "     is_representative_gene  \n",
       "76                    False  \n",
       "170                   False  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.representative_gene_df.query(\"ensembl_gene_id != ensembl_representative_gene_id\").head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d270f70f",
   "metadata": {
    "papermill": {
     "duration": 0.007924,
     "end_time": "2025-01-06T21:13:06.435110",
     "exception": false,
     "start_time": "2025-01-06T21:13:06.427186",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# replaced ID converter\n",
    "\n",
    "A single `old_stable_id` can map to multiple `new_stable_id`. For example, `ENSG00000152006`\n",
    "\n",
    "https://uswest.ensembl.org/Homo_sapiens/Tools/IDMapper/Results?tl=AzhM62SpkvdiLC4H-6808613\n",
    "\n",
    "Requested ID | Matched ID(s) | Releases\n",
    "-- | -- | --\n",
    "ENSG00000152006 | ENSG00000196273 | 26: ENSG00000196273.1\n",
    "ENSG00000152006 | ENSG00000197016 | 26: ENSG00000197016.1\n",
    "ENSG00000152006 | ENSG00000196239 | 26: ENSG00000196239.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3ec5dbf4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:06.452485Z",
     "iopub.status.busy": "2025-01-06T21:13:06.452019Z",
     "iopub.status.idle": "2025-01-06T21:13:13.925382Z",
     "shell.execute_reply": "2025-01-06T21:13:13.924658Z"
    },
    "papermill": {
     "duration": 7.483518,
     "end_time": "2025-01-06T21:13:13.926749",
     "exception": false,
     "start_time": "2025-01-06T21:13:06.443231",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>old_ensembl_gene_id</th>\n",
       "      <th>new_ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000037155</td>\n",
       "      <td>ENSMUSG00000044465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000700</td>\n",
       "      <td>ENSMUSG00000066358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  old_ensembl_gene_id new_ensembl_gene_id\n",
       "0  ENSMUSG00000037155  ENSMUSG00000044465\n",
       "1  ENSMUSG00000000700  ENSMUSG00000066358"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.old_to_new_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fca33efc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:13.944881Z",
     "iopub.status.busy": "2025-01-06T21:13:13.944477Z",
     "iopub.status.idle": "2025-01-06T21:13:13.966374Z",
     "shell.execute_reply": "2025-01-06T21:13:13.965853Z"
    },
    "papermill": {
     "duration": 0.032096,
     "end_time": "2025-01-06T21:13:13.967695",
     "exception": false,
     "start_time": "2025-01-06T21:13:13.935599",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSMUSG00000070062    131\n",
       "ENSMUSG00000070181    131\n",
       "Name: old_ensembl_gene_id, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# some ensembl genes replaced by many new ensembl genes\n",
    "ensg.old_to_new_df.old_ensembl_gene_id.value_counts().head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "578167ba",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:13.985249Z",
     "iopub.status.busy": "2025-01-06T21:13:13.985010Z",
     "iopub.status.idle": "2025-01-06T21:13:14.086012Z",
     "shell.execute_reply": "2025-01-06T21:13:14.085406Z"
    },
    "papermill": {
     "duration": 0.111308,
     "end_time": "2025-01-06T21:13:14.087318",
     "exception": false,
     "start_time": "2025-01-06T21:13:13.976010",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000152006'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# example\n",
    "ensg._update_ensembl_gene(\"ENSG00000152006\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "60afe952",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:14.105991Z",
     "iopub.status.busy": "2025-01-06T21:13:14.105568Z",
     "iopub.status.idle": "2025-01-06T21:13:14.822027Z",
     "shell.execute_reply": "2025-01-06T21:13:14.821349Z"
    },
    "papermill": {
     "duration": 0.727061,
     "end_time": "2025-01-06T21:13:14.823416",
     "exception": false,
     "start_time": "2025-01-06T21:13:14.096355",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>old_ensembl_gene_id</th>\n",
       "      <th>newest_ensembl_gene_id</th>\n",
       "      <th>is_current</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000700</td>\n",
       "      <td>ENSMUSG00000066358</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000700</td>\n",
       "      <td>ENSMUSG00000068941</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  old_ensembl_gene_id newest_ensembl_gene_id  is_current\n",
       "0  ENSMUSG00000000700     ENSMUSG00000066358       False\n",
       "1  ENSMUSG00000000700     ENSMUSG00000068941       False"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.old_to_newest_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "87fbbffe",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:14.841895Z",
     "iopub.status.busy": "2025-01-06T21:13:14.841703Z",
     "iopub.status.idle": "2025-01-06T21:13:14.856416Z",
     "shell.execute_reply": "2025-01-06T21:13:14.855787Z"
    },
    "papermill": {
     "duration": 0.025312,
     "end_time": "2025-01-06T21:13:14.857821",
     "exception": false,
     "start_time": "2025-01-06T21:13:14.832509",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71734"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ensg.old_to_newest_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d4944d3e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:14.876306Z",
     "iopub.status.busy": "2025-01-06T21:13:14.875915Z",
     "iopub.status.idle": "2025-01-06T21:13:14.891757Z",
     "shell.execute_reply": "2025-01-06T21:13:14.891246Z"
    },
    "papermill": {
     "duration": 0.026513,
     "end_time": "2025-01-06T21:13:14.893082",
     "exception": false,
     "start_time": "2025-01-06T21:13:14.866569",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     58891\n",
       "False    12843\n",
       "Name: is_current, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.old_to_newest_df.is_current.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ca74d1f",
   "metadata": {
    "papermill": {
     "duration": 0.008451,
     "end_time": "2025-01-06T21:13:14.910213",
     "exception": false,
     "start_time": "2025-01-06T21:13:14.901762",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## omni-updater\n",
    "\n",
    "The omni-updater dataset is designed to convert ensembl gene IDs from input data to the current, representative ensembl_gene_ids for this ensembl release. It assumes:\n",
    "\n",
    "- users want to update outdated genes with their replacements\n",
    "- users want a dataset of representative genes only, and want to convert alternative alleles to representative genes\n",
    "\n",
    "An inner join of a dataset with `update_df` on `input_ensembl_gene_id` will do the following:\n",
    "\n",
    "- produce output ensembl_gene_ids that are current and representatives\n",
    "- update outdated genes with their current identifiers. Outdated genes with no current replacement will be removed by the inner join.\n",
    "- update alternative gene alleles with their representatives\n",
    "- genes that are already represenative and current will map to themselves"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b3d54482",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:14.928425Z",
     "iopub.status.busy": "2025-01-06T21:13:14.928038Z",
     "iopub.status.idle": "2025-01-06T21:13:15.278664Z",
     "shell.execute_reply": "2025-01-06T21:13:15.277934Z"
    },
    "papermill": {
     "duration": 0.361264,
     "end_time": "2025-01-06T21:13:15.279974",
     "exception": false,
     "start_time": "2025-01-06T21:13:14.918710",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input_ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>input_current</th>\n",
       "      <th>input_representative</th>\n",
       "      <th>input_maps_to_n_genes</th>\n",
       "      <th>n_inputs_map_to_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  input_ensembl_gene_id     ensembl_gene_id  input_current  \\\n",
       "0    ENSMUSG00000000001  ENSMUSG00000000001           True   \n",
       "1    ENSMUSG00000000003  ENSMUSG00000000003           True   \n",
       "\n",
       "   input_representative  input_maps_to_n_genes  n_inputs_map_to_gene  \n",
       "0                  True                      1                     1  \n",
       "1                  True                      1                     1  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.update_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cb29e6da",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:15.299483Z",
     "iopub.status.busy": "2025-01-06T21:13:15.299047Z",
     "iopub.status.idle": "2025-01-06T21:13:15.326878Z",
     "shell.execute_reply": "2025-01-06T21:13:15.326210Z"
    },
    "papermill": {
     "duration": 0.038711,
     "end_time": "2025-01-06T21:13:15.328166",
     "exception": false,
     "start_time": "2025-01-06T21:13:15.289455",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input_ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>input_current</th>\n",
       "      <th>input_representative</th>\n",
       "      <th>input_maps_to_n_genes</th>\n",
       "      <th>n_inputs_map_to_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32001</th>\n",
       "      <td>ENSMUSG00000075866</td>\n",
       "      <td>ENSMUSG00000119351</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>126</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54448</th>\n",
       "      <td>ENSMUSG00000075866</td>\n",
       "      <td>ENSMUSG00000119496</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>126</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      input_ensembl_gene_id     ensembl_gene_id  input_current  \\\n",
       "32001    ENSMUSG00000075866  ENSMUSG00000119351          False   \n",
       "54448    ENSMUSG00000075866  ENSMUSG00000119496          False   \n",
       "\n",
       "       input_representative  input_maps_to_n_genes  n_inputs_map_to_gene  \n",
       "32001                  True                    126                   253  \n",
       "54448                  True                    126                    15  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.update_df.sort_values(\"input_maps_to_n_genes\", ascending=False).head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "693cde8a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:15.347881Z",
     "iopub.status.busy": "2025-01-06T21:13:15.347434Z",
     "iopub.status.idle": "2025-01-06T21:13:15.375837Z",
     "shell.execute_reply": "2025-01-06T21:13:15.375140Z"
    },
    "papermill": {
     "duration": 0.039512,
     "end_time": "2025-01-06T21:13:15.377255",
     "exception": false,
     "start_time": "2025-01-06T21:13:15.337743",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input_ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>input_current</th>\n",
       "      <th>input_representative</th>\n",
       "      <th>input_maps_to_n_genes</th>\n",
       "      <th>n_inputs_map_to_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>39819</th>\n",
       "      <td>ENSMUSG00000076000</td>\n",
       "      <td>ENSMUSG00000119739</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>120</td>\n",
       "      <td>255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44002</th>\n",
       "      <td>ENSMUSG00000075977</td>\n",
       "      <td>ENSMUSG00000119864</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>126</td>\n",
       "      <td>255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      input_ensembl_gene_id     ensembl_gene_id  input_current  \\\n",
       "39819    ENSMUSG00000076000  ENSMUSG00000119739          False   \n",
       "44002    ENSMUSG00000075977  ENSMUSG00000119864          False   \n",
       "\n",
       "       input_representative  input_maps_to_n_genes  n_inputs_map_to_gene  \n",
       "39819                  True                    120                   255  \n",
       "44002                  True                    126                   255  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.update_df.sort_values(\"n_inputs_map_to_gene\", ascending=False).head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "02474f5d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:15.397336Z",
     "iopub.status.busy": "2025-01-06T21:13:15.397133Z",
     "iopub.status.idle": "2025-01-06T21:13:15.413251Z",
     "shell.execute_reply": "2025-01-06T21:13:15.412699Z"
    },
    "papermill": {
     "duration": 0.027512,
     "end_time": "2025-01-06T21:13:15.414519",
     "exception": false,
     "start_time": "2025-01-06T21:13:15.387007",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5804328335362164"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(ensg.update_df.input_maps_to_n_genes == 1).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d7857d9d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:15.434157Z",
     "iopub.status.busy": "2025-01-06T21:13:15.433780Z",
     "iopub.status.idle": "2025-01-06T21:13:15.456970Z",
     "shell.execute_reply": "2025-01-06T21:13:15.456317Z"
    },
    "papermill": {
     "duration": 0.034464,
     "end_time": "2025-01-06T21:13:15.458351",
     "exception": false,
     "start_time": "2025-01-06T21:13:15.423887",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input_ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>input_current</th>\n",
       "      <th>input_representative</th>\n",
       "      <th>input_maps_to_n_genes</th>\n",
       "      <th>n_inputs_map_to_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [input_ensembl_gene_id, ensembl_gene_id, input_current, input_representative, input_maps_to_n_genes, n_inputs_map_to_gene]\n",
       "Index: []"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.update_df.query(\"ensembl_gene_id == 'ENSG00000256263'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1f0d8ae7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:15.478723Z",
     "iopub.status.busy": "2025-01-06T21:13:15.478312Z",
     "iopub.status.idle": "2025-01-06T21:13:15.519028Z",
     "shell.execute_reply": "2025-01-06T21:13:15.518356Z"
    },
    "papermill": {
     "duration": 0.052154,
     "end_time": "2025-01-06T21:13:15.520422",
     "exception": false,
     "start_time": "2025-01-06T21:13:15.468268",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The omni-updater contains 137,189 rows for mapping 82,538 input genes to 77,981 current, representative genes.\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    f\"The omni-updater contains {len(ensg.update_df):,} rows for mapping \"\n",
    "    f\"{ensg.update_df.input_ensembl_gene_id.nunique():,} input genes to \"\n",
    "    f\"{ensg.update_df.ensembl_gene_id.nunique():,} current, representative genes.\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8fae39d1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:15.541194Z",
     "iopub.status.busy": "2025-01-06T21:13:15.540746Z",
     "iopub.status.idle": "2025-01-06T21:13:15.561624Z",
     "shell.execute_reply": "2025-01-06T21:13:15.560894Z"
    },
    "papermill": {
     "duration": 0.032408,
     "end_time": "2025-01-06T21:13:15.562937",
     "exception": false,
     "start_time": "2025-01-06T21:13:15.530529",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>input_ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>input_current</th>\n",
       "      <th>input_representative</th>\n",
       "      <th>input_maps_to_n_genes</th>\n",
       "      <th>n_inputs_map_to_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [input_ensembl_gene_id, ensembl_gene_id, input_current, input_representative, input_maps_to_n_genes, n_inputs_map_to_gene]\n",
       "Index: []"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# https://useast.ensembl.org/Homo_sapiens/Tools/IDMapper/Results?tl=P45VLMbogubpI0QA-6815464\n",
    "ensg.update_df.query(\"input_ensembl_gene_id == 'ENSG00000201456'\").head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aa717e8",
   "metadata": {
    "papermill": {
     "duration": 0.009348,
     "end_time": "2025-01-06T21:13:15.581953",
     "exception": false,
     "start_time": "2025-01-06T21:13:15.572605",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## cross-refrences (xrefs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "af93fa0f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:13:15.601927Z",
     "iopub.status.busy": "2025-01-06T21:13:15.601474Z",
     "iopub.status.idle": "2025-01-06T21:15:58.018766Z",
     "shell.execute_reply": "2025-01-06T21:15:58.018050Z"
    },
    "papermill": {
     "duration": 162.438815,
     "end_time": "2025-01-06T21:15:58.030143",
     "exception": false,
     "start_time": "2025-01-06T21:13:15.591328",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_representative_gene_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>xref_source</th>\n",
       "      <th>xref_accession</th>\n",
       "      <th>xref_label</th>\n",
       "      <th>xref_description</th>\n",
       "      <th>xref_info_type</th>\n",
       "      <th>xref_linkage_annotation</th>\n",
       "      <th>xref_curie</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>ArrayExpress</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>None</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>None</td>\n",
       "      <td>arrayexpress:ENSMUSG00000000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>EntrezGene</td>\n",
       "      <td>14679</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>G protein subunit alpha i3</td>\n",
       "      <td>DEPENDENT</td>\n",
       "      <td>None</td>\n",
       "      <td>ncbigene:14679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:95773</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>G protein subunit alpha i3</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>None</td>\n",
       "      <td>mgi:95773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>Reactome_gene</td>\n",
       "      <td>R-MMU-109582</td>\n",
       "      <td>R-MMU-109582</td>\n",
       "      <td>Hemostasis</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>None</td>\n",
       "      <td>reactome:R-MMU-109582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>Reactome_gene</td>\n",
       "      <td>R-MMU-111885</td>\n",
       "      <td>R-MMU-111885</td>\n",
       "      <td>Opioid Signalling</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>None</td>\n",
       "      <td>reactome:R-MMU-111885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ensembl_representative_gene_id     ensembl_gene_id gene_symbol  \\\n",
       "0             ENSMUSG00000000001  ENSMUSG00000000001       Gnai3   \n",
       "1             ENSMUSG00000000001  ENSMUSG00000000001       Gnai3   \n",
       "2             ENSMUSG00000000001  ENSMUSG00000000001       Gnai3   \n",
       "3             ENSMUSG00000000001  ENSMUSG00000000001       Gnai3   \n",
       "4             ENSMUSG00000000001  ENSMUSG00000000001       Gnai3   \n",
       "\n",
       "     xref_source      xref_accession          xref_label  \\\n",
       "0   ArrayExpress  ENSMUSG00000000001  ENSMUSG00000000001   \n",
       "1     EntrezGene               14679               Gnai3   \n",
       "2            MGI           MGI:95773               Gnai3   \n",
       "3  Reactome_gene        R-MMU-109582        R-MMU-109582   \n",
       "4  Reactome_gene        R-MMU-111885        R-MMU-111885   \n",
       "\n",
       "             xref_description xref_info_type xref_linkage_annotation  \\\n",
       "0                        None         DIRECT                    None   \n",
       "1  G protein subunit alpha i3      DEPENDENT                    None   \n",
       "2  G protein subunit alpha i3         DIRECT                    None   \n",
       "3                  Hemostasis         DIRECT                    None   \n",
       "4           Opioid Signalling         DIRECT                    None   \n",
       "\n",
       "                        xref_curie  \n",
       "0  arrayexpress:ENSMUSG00000000001  \n",
       "1                   ncbigene:14679  \n",
       "2                        mgi:95773  \n",
       "3            reactome:R-MMU-109582  \n",
       "4            reactome:R-MMU-111885  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.xref_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ad047394",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:15:58.050687Z",
     "iopub.status.busy": "2025-01-06T21:15:58.050269Z",
     "iopub.status.idle": "2025-01-06T21:15:58.257001Z",
     "shell.execute_reply": "2025-01-06T21:15:58.256270Z"
    },
    "papermill": {
     "duration": 0.218571,
     "end_time": "2025-01-06T21:15:58.258397",
     "exception": false,
     "start_time": "2025-01-06T21:15:58.039826",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: xref_source, dtype: int64)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# datasets where there are ensembl_gene_id-xref_source-xref_accession pairs might not be distinct \n",
    "xref_dup_df = ensg.xref_df[ensg.xref_df.duplicated(subset=[\"ensembl_gene_id\", \"xref_source\", \"xref_accession\"], keep=False)]\n",
    "xref_dup_df.xref_source.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "aa7799f4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:15:58.280361Z",
     "iopub.status.busy": "2025-01-06T21:15:58.279933Z",
     "iopub.status.idle": "2025-01-06T21:15:58.585345Z",
     "shell.execute_reply": "2025-01-06T21:15:58.584645Z"
    },
    "papermill": {
     "duration": 0.317647,
     "end_time": "2025-01-06T21:15:58.586855",
     "exception": false,
     "start_time": "2025-01-06T21:15:58.269208",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>xref_info_type</th>\n",
       "      <th>DEPENDENT</th>\n",
       "      <th>DIRECT</th>\n",
       "      <th>MISC</th>\n",
       "      <th>PROJECTION</th>\n",
       "      <th>All</th>\n",
       "      <th>bioregistry_prefix</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xref_source</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ArrayExpress</th>\n",
       "      <td>0</td>\n",
       "      <td>78298</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>78298</td>\n",
       "      <td>arrayexpress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EntrezGene</th>\n",
       "      <td>28531</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28531</td>\n",
       "      <td>ncbigene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>hgnc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGI</th>\n",
       "      <td>0</td>\n",
       "      <td>56128</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56128</td>\n",
       "      <td>mgi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RFAM</th>\n",
       "      <td>0</td>\n",
       "      <td>2467</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2467</td>\n",
       "      <td>rfam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reactome_gene</th>\n",
       "      <td>0</td>\n",
       "      <td>88274</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88274</td>\n",
       "      <td>reactome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot_gn</th>\n",
       "      <td>51465</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>51465</td>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WikiGene</th>\n",
       "      <td>28531</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28531</td>\n",
       "      <td>wikigenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>miRBase</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1360</td>\n",
       "      <td>0</td>\n",
       "      <td>1360</td>\n",
       "      <td>mirbase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>108527</td>\n",
       "      <td>225167</td>\n",
       "      <td>1360</td>\n",
       "      <td>3</td>\n",
       "      <td>335057</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "xref_info_type  DEPENDENT  DIRECT  MISC  PROJECTION     All bioregistry_prefix\n",
       "xref_source                                                                   \n",
       "ArrayExpress            0   78298     0           0   78298       arrayexpress\n",
       "EntrezGene          28531       0     0           0   28531           ncbigene\n",
       "HGNC                    0       0     0           3       3               hgnc\n",
       "MGI                     0   56128     0           0   56128                mgi\n",
       "RFAM                    0    2467     0           0    2467               rfam\n",
       "Reactome_gene           0   88274     0           0   88274           reactome\n",
       "Uniprot_gn          51465       0     0           0   51465            uniprot\n",
       "WikiGene            28531       0     0           0   28531          wikigenes\n",
       "miRBase                 0       0  1360           0    1360            mirbase\n",
       "All                108527  225167  1360           3  335057               None"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# xref sources versus info_types\n",
    "df = pd.crosstab(ensg.xref_df.xref_source, ensg.xref_df.xref_info_type, margins=True)\n",
    "df[\"bioregistry_prefix\"] = df.index.to_series().replace(ensg._xref_prefix_updater).map(normalize_prefix)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5116bcd",
   "metadata": {
    "papermill": {
     "duration": 0.009933,
     "end_time": "2025-01-06T21:15:58.607637",
     "exception": false,
     "start_time": "2025-01-06T21:15:58.597704",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## Gene Ontology xrefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ac5d62ca",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:15:58.628954Z",
     "iopub.status.busy": "2025-01-06T21:15:58.628687Z",
     "iopub.status.idle": "2025-01-06T21:16:32.801535Z",
     "shell.execute_reply": "2025-01-06T21:16:32.800817Z"
    },
    "papermill": {
     "duration": 34.195833,
     "end_time": "2025-01-06T21:16:32.813452",
     "exception": false,
     "start_time": "2025-01-06T21:15:58.617619",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>go_id</th>\n",
       "      <th>go_label</th>\n",
       "      <th>go_evidence_codes</th>\n",
       "      <th>xref_info_types</th>\n",
       "      <th>xref_info_texts</th>\n",
       "      <th>ensembl_transcript_ids</th>\n",
       "      <th>ensembl_representative_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>GO:0000139</td>\n",
       "      <td>Golgi membrane</td>\n",
       "      <td>IEA</td>\n",
       "      <td>PROJECTION</td>\n",
       "      <td>from homo_sapiens translation ENSP00000358867</td>\n",
       "      <td>ENSMUST00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>GO:0001664</td>\n",
       "      <td>G protein-coupled receptor binding</td>\n",
       "      <td>IBA</td>\n",
       "      <td>DIRECT</td>\n",
       "      <td>GO_Central</td>\n",
       "      <td>ENSMUST00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>GO:0003924</td>\n",
       "      <td>GTPase activity</td>\n",
       "      <td>IBA,IEA,ISS,TAS</td>\n",
       "      <td>PROJECTION,DEPENDENT,DIRECT</td>\n",
       "      <td>,from homo_sapiens translation ENSP00000358867...</td>\n",
       "      <td>ENSMUST00000000001</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ensembl_gene_id       go_id                            go_label  \\\n",
       "0  ENSMUSG00000000001  GO:0000139                      Golgi membrane   \n",
       "1  ENSMUSG00000000001  GO:0001664  G protein-coupled receptor binding   \n",
       "2  ENSMUSG00000000001  GO:0003924                     GTPase activity   \n",
       "\n",
       "  go_evidence_codes              xref_info_types  \\\n",
       "0               IEA                   PROJECTION   \n",
       "1               IBA                       DIRECT   \n",
       "2   IBA,IEA,ISS,TAS  PROJECTION,DEPENDENT,DIRECT   \n",
       "\n",
       "                                     xref_info_texts ensembl_transcript_ids  \\\n",
       "0      from homo_sapiens translation ENSP00000358867     ENSMUST00000000001   \n",
       "1                                         GO_Central     ENSMUST00000000001   \n",
       "2  ,from homo_sapiens translation ENSP00000358867...     ENSMUST00000000001   \n",
       "\n",
       "  ensembl_representative_gene_id  \n",
       "0             ENSMUSG00000000001  \n",
       "1             ENSMUSG00000000001  \n",
       "2             ENSMUSG00000000001  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.xref_go_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8dac6754",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:32.835029Z",
     "iopub.status.busy": "2025-01-06T21:16:32.834587Z",
     "iopub.status.idle": "2025-01-06T21:16:32.860728Z",
     "shell.execute_reply": "2025-01-06T21:16:32.860015Z"
    },
    "papermill": {
     "duration": 0.038496,
     "end_time": "2025-01-06T21:16:32.862085",
     "exception": false,
     "start_time": "2025-01-06T21:16:32.823589",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GO terms for CCR5\n",
    "# compare to http://useast.ensembl.org/Homo_sapiens/Gene/Ontologies/molecular_function?g=ENSG00000160791\n",
    "sorted(ensg.xref_go_df.query(\"ensembl_gene_id == 'ENSG00000160791'\").go_label)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a8ebb18",
   "metadata": {
    "papermill": {
     "duration": 0.010424,
     "end_time": "2025-01-06T21:16:32.883486",
     "exception": false,
     "start_time": "2025-01-06T21:16:32.873062",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## lrg xrefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1d6bd7af",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:32.905488Z",
     "iopub.status.busy": "2025-01-06T21:16:32.904943Z",
     "iopub.status.idle": "2025-01-06T21:16:32.922006Z",
     "shell.execute_reply": "2025-01-06T21:16:32.921331Z"
    },
    "papermill": {
     "duration": 0.02967,
     "end_time": "2025-01-06T21:16:32.923336",
     "exception": false,
     "start_time": "2025-01-06T21:16:32.893666",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>lrg_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ensembl_gene_id, lrg_gene_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.xref_lrg_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a7634059",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:32.946036Z",
     "iopub.status.busy": "2025-01-06T21:16:32.945481Z",
     "iopub.status.idle": "2025-01-06T21:16:32.960399Z",
     "shell.execute_reply": "2025-01-06T21:16:32.959849Z"
    },
    "papermill": {
     "duration": 0.027504,
     "end_time": "2025-01-06T21:16:32.961747",
     "exception": false,
     "start_time": "2025-01-06T21:16:32.934243",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ensg.xref_lrg_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d71d26cf",
   "metadata": {
    "papermill": {
     "duration": 0.010244,
     "end_time": "2025-01-06T21:16:32.982776",
     "exception": false,
     "start_time": "2025-01-06T21:16:32.972532",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### ncbigene xrefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "82157d02",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:33.004639Z",
     "iopub.status.busy": "2025-01-06T21:16:33.004136Z",
     "iopub.status.idle": "2025-01-06T21:16:33.091667Z",
     "shell.execute_reply": "2025-01-06T21:16:33.090942Z"
    },
    "papermill": {
     "duration": 0.100258,
     "end_time": "2025-01-06T21:16:33.093215",
     "exception": false,
     "start_time": "2025-01-06T21:16:32.992957",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_representative_gene_id</th>\n",
       "      <th>ncbigene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>ncbigene_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>14679</td>\n",
       "      <td>Gnai3</td>\n",
       "      <td>Gnai3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "      <td>54192</td>\n",
       "      <td>Pbsn</td>\n",
       "      <td>Pbsn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENSMUSG00000000028</td>\n",
       "      <td>12544</td>\n",
       "      <td>Cdc45</td>\n",
       "      <td>Cdc45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ENSMUSG00000000031</td>\n",
       "      <td>14955</td>\n",
       "      <td>H19</td>\n",
       "      <td>H19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>107815</td>\n",
       "      <td>Scml2</td>\n",
       "      <td>Scml2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ensembl_representative_gene_id ncbigene_id gene_symbol ncbigene_symbol\n",
       "1              ENSMUSG00000000001       14679       Gnai3           Gnai3\n",
       "26             ENSMUSG00000000003       54192        Pbsn            Pbsn\n",
       "33             ENSMUSG00000000028       12544       Cdc45           Cdc45\n",
       "51             ENSMUSG00000000031       14955         H19             H19\n",
       "55             ENSMUSG00000000037      107815       Scml2           Scml2"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensg.xref_ncbigene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a602d412",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:33.117049Z",
     "iopub.status.busy": "2025-01-06T21:16:33.116528Z",
     "iopub.status.idle": "2025-01-06T21:16:33.141603Z",
     "shell.execute_reply": "2025-01-06T21:16:33.141000Z"
    },
    "papermill": {
     "duration": 0.038023,
     "end_time": "2025-01-06T21:16:33.142925",
     "exception": false,
     "start_time": "2025-01-06T21:16:33.104902",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSMUSG00000000001    1\n",
       "ENSMUSG00000071531    1\n",
       "ENSMUSG00000071573    1\n",
       "Name: ensembl_representative_gene_id, dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ensembl gene mapped to by multiple ncbigenes\n",
    "ensg.xref_ncbigene_df.ensembl_representative_gene_id.value_counts().head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "957dfeb0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:33.166462Z",
     "iopub.status.busy": "2025-01-06T21:16:33.165890Z",
     "iopub.status.idle": "2025-01-06T21:16:33.183997Z",
     "shell.execute_reply": "2025-01-06T21:16:33.183396Z"
    },
    "papermill": {
     "duration": 0.03108,
     "end_time": "2025-01-06T21:16:33.185382",
     "exception": false,
     "start_time": "2025-01-06T21:16:33.154302",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28204, 0)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ensg.xref_ncbigene_df), ensg.xref_ncbigene_df.ensembl_representative_gene_id.duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d1d9791d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:33.209574Z",
     "iopub.status.busy": "2025-01-06T21:16:33.209107Z",
     "iopub.status.idle": "2025-01-06T21:16:33.235640Z",
     "shell.execute_reply": "2025-01-06T21:16:33.234986Z"
    },
    "papermill": {
     "duration": 0.040083,
     "end_time": "2025-01-06T21:16:33.237134",
     "exception": false,
     "start_time": "2025-01-06T21:16:33.197051",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "108168560    7\n",
       "102635519    4\n",
       "22301        2\n",
       "Name: ncbigene_id, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ncbigene mapped to by multiple ensembl genes, likely due to alt gene alleles\n",
    "ensg.xref_ncbigene_df.ncbigene_id.value_counts().head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "efcd49b0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:33.261329Z",
     "iopub.status.busy": "2025-01-06T21:16:33.260795Z",
     "iopub.status.idle": "2025-01-06T21:16:33.279328Z",
     "shell.execute_reply": "2025-01-06T21:16:33.278766Z"
    },
    "papermill": {
     "duration": 0.031747,
     "end_time": "2025-01-06T21:16:33.280694",
     "exception": false,
     "start_time": "2025-01-06T21:16:33.248947",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28204, 28)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ensg.xref_ncbigene_df), ensg.xref_ncbigene_df.ncbigene_id.duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "1226814a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:33.304938Z",
     "iopub.status.busy": "2025-01-06T21:16:33.304411Z",
     "iopub.status.idle": "2025-01-06T21:16:33.318665Z",
     "shell.execute_reply": "2025-01-06T21:16:33.318129Z"
    },
    "papermill": {
     "duration": 0.027777,
     "end_time": "2025-01-06T21:16:33.320086",
     "exception": false,
     "start_time": "2025-01-06T21:16:33.292309",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ensg.xref_ncbigene_df.query(\"ensembl_representative_gene_id == 'ENSG00000231500'\")\n",
    "# ensg.xref_ncbigene_df.query(\"ncbigene_id == '51206'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "76cbfef4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:33.344238Z",
     "iopub.status.busy": "2025-01-06T21:16:33.343708Z",
     "iopub.status.idle": "2025-01-06T21:16:33.371825Z",
     "shell.execute_reply": "2025-01-06T21:16:33.371189Z"
    },
    "papermill": {
     "duration": 0.041344,
     "end_time": "2025-01-06T21:16:33.373145",
     "exception": false,
     "start_time": "2025-01-06T21:16:33.331801",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "77981"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repr_ensembl_gene_ids = set(ensg.gene_df.ensembl_representative_gene_id)\n",
    "len(repr_ensembl_gene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f514b721",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-06T21:16:33.397559Z",
     "iopub.status.busy": "2025-01-06T21:16:33.397014Z",
     "iopub.status.idle": "2025-01-06T21:16:33.443358Z",
     "shell.execute_reply": "2025-01-06T21:16:33.442663Z"
    },
    "papermill": {
     "duration": 0.059666,
     "end_time": "2025-01-06T21:16:33.444725",
     "exception": false,
     "start_time": "2025-01-06T21:16:33.385059",
     "status": "completed"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_gene_version</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_symbol_source_db</th>\n",
       "      <th>gene_symbol_source_id</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>ensembl_source</th>\n",
       "      <th>ensembl_created_date</th>\n",
       "      <th>ensembl_modified_date</th>\n",
       "      <th>coord_system_version</th>\n",
       "      <th>...</th>\n",
       "      <th>seq_region_start</th>\n",
       "      <th>seq_region_end</th>\n",
       "      <th>seq_region_strand</th>\n",
       "      <th>primary_assembly</th>\n",
       "      <th>lrg_gene_id</th>\n",
       "      <th>mhc</th>\n",
       "      <th>gene_description</th>\n",
       "      <th>gene_description_source_db</th>\n",
       "      <th>gene_description_source_id</th>\n",
       "      <th>ensembl_representative_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17497</th>\n",
       "      <td>ENSMUSG00000062783</td>\n",
       "      <td>3</td>\n",
       "      <td>Csprs</td>\n",
       "      <td>EntrezGene</td>\n",
       "      <td>114564</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:28:04</td>\n",
       "      <td>2020-08-25 18:37:28</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>111572</td>\n",
       "      <td>163011</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>component of Sp100-rs</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>114564</td>\n",
       "      <td>ENSMUSG00000062783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20614</th>\n",
       "      <td>ENSMUSG00000074720</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000074720</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:31:13</td>\n",
       "      <td>2020-08-25 18:37:19</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>25923</td>\n",
       "      <td>27230</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000074720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22214</th>\n",
       "      <td>ENSMUSG00000079190</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000079190</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:17:40</td>\n",
       "      <td>2020-08-25 18:37:25</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>167446</td>\n",
       "      <td>196478</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000079190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22215</th>\n",
       "      <td>ENSMUSG00000079192</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000079192</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:11:51</td>\n",
       "      <td>2020-08-25 18:37:27</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>123792</td>\n",
       "      <td>124928</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000079192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22218</th>\n",
       "      <td>ENSMUSG00000079222</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000079222</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:26:11</td>\n",
       "      <td>2020-08-25 18:37:27</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>17591</td>\n",
       "      <td>30203</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000079222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22441</th>\n",
       "      <td>ENSMUSG00000079794</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000079794</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:13:32</td>\n",
       "      <td>2020-08-25 18:37:27</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>135395</td>\n",
       "      <td>136519</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000079794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22442</th>\n",
       "      <td>ENSMUSG00000079800</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000079800</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:09:17</td>\n",
       "      <td>2020-08-25 18:37:26</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>9124</td>\n",
       "      <td>58882</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000079800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22444</th>\n",
       "      <td>ENSMUSG00000079808</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000079808</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:05:55</td>\n",
       "      <td>2020-08-25 18:37:18</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>128555</td>\n",
       "      <td>150452</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000079808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30066</th>\n",
       "      <td>ENSMUSG00000091585</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000091585</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:39:52</td>\n",
       "      <td>2020-08-25 18:37:22</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>716296</td>\n",
       "      <td>720517</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000091585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31226</th>\n",
       "      <td>ENSMUSG00000094054</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094054</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:15:58</td>\n",
       "      <td>2020-08-25 18:37:25</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>113868</td>\n",
       "      <td>114758</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31286</th>\n",
       "      <td>ENSMUSG00000094172</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094172</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:46:01</td>\n",
       "      <td>2020-08-25 18:37:22</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>562161</td>\n",
       "      <td>564894</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31359</th>\n",
       "      <td>ENSMUSG00000094303</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094303</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 15:58:15</td>\n",
       "      <td>2020-08-25 18:37:19</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>19740</td>\n",
       "      <td>22248</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31379</th>\n",
       "      <td>ENSMUSG00000094337</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSMUSG00000094337</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:32:12</td>\n",
       "      <td>2020-08-25 18:37:19</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>82993</td>\n",
       "      <td>86051</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31399</th>\n",
       "      <td>ENSMUSG00000094383</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094383</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:38:10</td>\n",
       "      <td>2020-08-25 18:37:21</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>90482</td>\n",
       "      <td>93220</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31439</th>\n",
       "      <td>ENSMUSG00000094474</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094474</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:44:31</td>\n",
       "      <td>2020-08-25 18:37:21</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>173050</td>\n",
       "      <td>176134</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31460</th>\n",
       "      <td>ENSMUSG00000094514</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094514</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:18:46</td>\n",
       "      <td>2020-08-25 18:37:26</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>226405</td>\n",
       "      <td>229375</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31569</th>\n",
       "      <td>ENSMUSG00000094722</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094722</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 15:53:54</td>\n",
       "      <td>2020-08-25 18:37:24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>90528</td>\n",
       "      <td>93623</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31572</th>\n",
       "      <td>ENSMUSG00000094728</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094728</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:27:52</td>\n",
       "      <td>2020-08-25 18:37:28</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>58743</td>\n",
       "      <td>59114</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31581</th>\n",
       "      <td>ENSMUSG00000094741</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094741</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:28:44</td>\n",
       "      <td>2020-08-25 18:37:18</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>6366</td>\n",
       "      <td>8939</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31610</th>\n",
       "      <td>ENSMUSG00000094791</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094791</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:49:25</td>\n",
       "      <td>2020-08-25 18:37:21</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>314841</td>\n",
       "      <td>317584</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31616</th>\n",
       "      <td>ENSMUSG00000094799</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094799</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:15:52</td>\n",
       "      <td>2020-08-25 18:37:27</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>147792</td>\n",
       "      <td>149707</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31635</th>\n",
       "      <td>ENSMUSG00000094836</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094836</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:30:28</td>\n",
       "      <td>2020-08-25 18:37:21</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>4975</td>\n",
       "      <td>7356</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31645</th>\n",
       "      <td>ENSMUSG00000094855</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094855</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:40:26</td>\n",
       "      <td>2020-08-25 18:37:23</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>921942</td>\n",
       "      <td>924675</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31656</th>\n",
       "      <td>ENSMUSG00000094874</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094874</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:27:03</td>\n",
       "      <td>2020-08-25 18:37:28</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>30394</td>\n",
       "      <td>32306</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31666</th>\n",
       "      <td>ENSMUSG00000094887</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094887</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:44:31</td>\n",
       "      <td>2020-08-25 18:37:22</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>587624</td>\n",
       "      <td>625594</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31678</th>\n",
       "      <td>ENSMUSG00000094915</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000094915</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:04:14</td>\n",
       "      <td>2020-08-25 18:37:18</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>31967</td>\n",
       "      <td>34932</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000094915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31728</th>\n",
       "      <td>ENSMUSG00000095019</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095019</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:37:12</td>\n",
       "      <td>2020-08-25 18:37:24</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>81607</td>\n",
       "      <td>82689</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31738</th>\n",
       "      <td>ENSMUSG00000095041</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSMUSG00000095041</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:39:20</td>\n",
       "      <td>2020-08-25 18:37:20</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>52190</td>\n",
       "      <td>59690</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31760</th>\n",
       "      <td>ENSMUSG00000095092</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095092</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:10:08</td>\n",
       "      <td>2020-08-25 18:37:27</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>108390</td>\n",
       "      <td>110303</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31836</th>\n",
       "      <td>ENSMUSG00000095250</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095250</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:12:45</td>\n",
       "      <td>2020-08-25 18:37:25</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>30483</td>\n",
       "      <td>32396</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31940</th>\n",
       "      <td>ENSMUSG00000095450</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095450</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:27:03</td>\n",
       "      <td>2020-08-25 18:37:28</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>46295</td>\n",
       "      <td>46666</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31952</th>\n",
       "      <td>ENSMUSG00000095475</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095475</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:49:39</td>\n",
       "      <td>2020-08-25 18:37:23</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>910289</td>\n",
       "      <td>913083</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31966</th>\n",
       "      <td>ENSMUSG00000095500</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095500</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:27:03</td>\n",
       "      <td>2020-08-25 18:37:28</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>32239</td>\n",
       "      <td>47294</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31968</th>\n",
       "      <td>ENSMUSG00000095505</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095505</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:36:37</td>\n",
       "      <td>2020-08-25 18:37:29</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>195181</td>\n",
       "      <td>197903</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31979</th>\n",
       "      <td>ENSMUSG00000095523</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095523</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:48:06</td>\n",
       "      <td>2020-08-25 18:37:23</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>837364</td>\n",
       "      <td>840451</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32008</th>\n",
       "      <td>ENSMUSG00000095570</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095570</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:33:52</td>\n",
       "      <td>2020-08-25 18:37:19</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>146926</td>\n",
       "      <td>149659</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32069</th>\n",
       "      <td>ENSMUSG00000095666</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095666</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:06:55</td>\n",
       "      <td>2020-08-25 18:37:29</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>41338</td>\n",
       "      <td>44417</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32073</th>\n",
       "      <td>ENSMUSG00000095672</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095672</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:16:46</td>\n",
       "      <td>2020-08-25 18:37:25</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>160465</td>\n",
       "      <td>162404</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32112</th>\n",
       "      <td>ENSMUSG00000095763</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095763</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:38:53</td>\n",
       "      <td>2020-08-25 18:37:22</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>806210</td>\n",
       "      <td>809281</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32123</th>\n",
       "      <td>ENSMUSG00000095787</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000095787</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:14:23</td>\n",
       "      <td>2020-08-25 18:37:25</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>54469</td>\n",
       "      <td>54840</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000095787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32289</th>\n",
       "      <td>ENSMUSG00000096100</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000096100</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:15:05</td>\n",
       "      <td>2020-08-25 18:37:25</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>66955</td>\n",
       "      <td>67326</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000096100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32358</th>\n",
       "      <td>ENSMUSG00000096236</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000096236</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 15:59:52</td>\n",
       "      <td>2020-08-25 18:37:20</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>31101</td>\n",
       "      <td>33667</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000096236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32359</th>\n",
       "      <td>ENSMUSG00000096244</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000096244</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:38:48</td>\n",
       "      <td>2020-08-25 18:37:23</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>149261</td>\n",
       "      <td>150163</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>vomeronasal 1 receptor Vmn1r187</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>100039499</td>\n",
       "      <td>ENSMUSG00000096244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32509</th>\n",
       "      <td>ENSMUSG00000096550</td>\n",
       "      <td>2</td>\n",
       "      <td>Gm16367</td>\n",
       "      <td>EntrezGene</td>\n",
       "      <td>545762</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:49:11</td>\n",
       "      <td>2020-08-25 18:37:21</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>477246</td>\n",
       "      <td>480331</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>predicted gene 16367</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>545762</td>\n",
       "      <td>ENSMUSG00000096550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32572</th>\n",
       "      <td>ENSMUSG00000096680</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000096680</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:40:55</td>\n",
       "      <td>2020-08-25 18:37:29</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>130715</td>\n",
       "      <td>133448</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000096680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32606</th>\n",
       "      <td>ENSMUSG00000096728</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000096728</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:35:31</td>\n",
       "      <td>2020-08-25 18:37:19</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>158543</td>\n",
       "      <td>161339</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000096728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32626</th>\n",
       "      <td>ENSMUSG00000096756</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000096756</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:01:38</td>\n",
       "      <td>2020-08-25 18:37:20</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>108318</td>\n",
       "      <td>111392</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000096756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32639</th>\n",
       "      <td>ENSMUSG00000096776</td>\n",
       "      <td>2</td>\n",
       "      <td>Vmn1r186</td>\n",
       "      <td>EntrezGene</td>\n",
       "      <td>100039479</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 17:37:35</td>\n",
       "      <td>2020-08-25 18:37:23</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>94186</td>\n",
       "      <td>95088</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>vomeronasal 1 receptor Vmn1r186</td>\n",
       "      <td>NCBI gene (formerly Entrezgene)</td>\n",
       "      <td>100039479</td>\n",
       "      <td>ENSMUSG00000096776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32654</th>\n",
       "      <td>ENSMUSG00000096808</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000096808</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-04-27 16:28:04</td>\n",
       "      <td>2020-08-25 18:37:28</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>151235</td>\n",
       "      <td>152734</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000096808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76129</th>\n",
       "      <td>ENSMUSG00000144196</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000144196</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>havana_tagene</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>631</td>\n",
       "      <td>156515</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000144196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76130</th>\n",
       "      <td>ENSMUSG00000144199</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000144199</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>havana_tagene</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>36491</td>\n",
       "      <td>157386</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000144199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76131</th>\n",
       "      <td>ENSMUSG00000144201</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000144201</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>havana_tagene</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>24951</td>\n",
       "      <td>56818</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000144201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76132</th>\n",
       "      <td>ENSMUSG00000144203</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000144203</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>havana_tagene</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>147544</td>\n",
       "      <td>163165</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000144203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76133</th>\n",
       "      <td>ENSMUSG00000144204</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000144204</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>havana_tagene</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>2024-05-06 19:10:14</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>89064</td>\n",
       "      <td>165170</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00000144204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76204</th>\n",
       "      <td>ENSMUSG00002074899</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002074899</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>13262</td>\n",
       "      <td>13382</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00002074899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76275</th>\n",
       "      <td>ENSMUSG00002074970</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002074970</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>31243</td>\n",
       "      <td>31343</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00002074970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76705</th>\n",
       "      <td>ENSMUSG00002075400</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002075400</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>38438</td>\n",
       "      <td>38535</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00002075400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77034</th>\n",
       "      <td>ENSMUSG00002075729</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002075729</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>32719</td>\n",
       "      <td>32818</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00002075729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78195</th>\n",
       "      <td>ENSMUSG00002076890</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002076890</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>2020-12-10 05:01:07</td>\n",
       "      <td>GRCm39</td>\n",
       "      <td>...</td>\n",
       "      <td>16623</td>\n",
       "      <td>16721</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSG00002076890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ensembl_gene_id  ensembl_gene_version         gene_symbol  \\\n",
       "17497  ENSMUSG00000062783                     3               Csprs   \n",
       "20614  ENSMUSG00000074720                     2  ENSMUSG00000074720   \n",
       "22214  ENSMUSG00000079190                     4  ENSMUSG00000079190   \n",
       "22215  ENSMUSG00000079192                     3  ENSMUSG00000079192   \n",
       "22218  ENSMUSG00000079222                     3  ENSMUSG00000079222   \n",
       "22441  ENSMUSG00000079794                     3  ENSMUSG00000079794   \n",
       "22442  ENSMUSG00000079800                     3  ENSMUSG00000079800   \n",
       "22444  ENSMUSG00000079808                     4  ENSMUSG00000079808   \n",
       "30066  ENSMUSG00000091585                     3  ENSMUSG00000091585   \n",
       "31226  ENSMUSG00000094054                     2  ENSMUSG00000094054   \n",
       "31286  ENSMUSG00000094172                     2  ENSMUSG00000094172   \n",
       "31359  ENSMUSG00000094303                     2  ENSMUSG00000094303   \n",
       "31379  ENSMUSG00000094337                     8  ENSMUSG00000094337   \n",
       "31399  ENSMUSG00000094383                     2  ENSMUSG00000094383   \n",
       "31439  ENSMUSG00000094474                     2  ENSMUSG00000094474   \n",
       "31460  ENSMUSG00000094514                     2  ENSMUSG00000094514   \n",
       "31569  ENSMUSG00000094722                     2  ENSMUSG00000094722   \n",
       "31572  ENSMUSG00000094728                     2  ENSMUSG00000094728   \n",
       "31581  ENSMUSG00000094741                     2  ENSMUSG00000094741   \n",
       "31610  ENSMUSG00000094791                     2  ENSMUSG00000094791   \n",
       "31616  ENSMUSG00000094799                     2  ENSMUSG00000094799   \n",
       "31635  ENSMUSG00000094836                     2  ENSMUSG00000094836   \n",
       "31645  ENSMUSG00000094855                     2  ENSMUSG00000094855   \n",
       "31656  ENSMUSG00000094874                     2  ENSMUSG00000094874   \n",
       "31666  ENSMUSG00000094887                     2  ENSMUSG00000094887   \n",
       "31678  ENSMUSG00000094915                     2  ENSMUSG00000094915   \n",
       "31728  ENSMUSG00000095019                     2  ENSMUSG00000095019   \n",
       "31738  ENSMUSG00000095041                     8  ENSMUSG00000095041   \n",
       "31760  ENSMUSG00000095092                     2  ENSMUSG00000095092   \n",
       "31836  ENSMUSG00000095250                     2  ENSMUSG00000095250   \n",
       "31940  ENSMUSG00000095450                     2  ENSMUSG00000095450   \n",
       "31952  ENSMUSG00000095475                     2  ENSMUSG00000095475   \n",
       "31966  ENSMUSG00000095500                     2  ENSMUSG00000095500   \n",
       "31968  ENSMUSG00000095505                     2  ENSMUSG00000095505   \n",
       "31979  ENSMUSG00000095523                     2  ENSMUSG00000095523   \n",
       "32008  ENSMUSG00000095570                     2  ENSMUSG00000095570   \n",
       "32069  ENSMUSG00000095666                     2  ENSMUSG00000095666   \n",
       "32073  ENSMUSG00000095672                     2  ENSMUSG00000095672   \n",
       "32112  ENSMUSG00000095763                     2  ENSMUSG00000095763   \n",
       "32123  ENSMUSG00000095787                     2  ENSMUSG00000095787   \n",
       "32289  ENSMUSG00000096100                     2  ENSMUSG00000096100   \n",
       "32358  ENSMUSG00000096236                     2  ENSMUSG00000096236   \n",
       "32359  ENSMUSG00000096244                     2  ENSMUSG00000096244   \n",
       "32509  ENSMUSG00000096550                     2             Gm16367   \n",
       "32572  ENSMUSG00000096680                     2  ENSMUSG00000096680   \n",
       "32606  ENSMUSG00000096728                     2  ENSMUSG00000096728   \n",
       "32626  ENSMUSG00000096756                     2  ENSMUSG00000096756   \n",
       "32639  ENSMUSG00000096776                     2            Vmn1r186   \n",
       "32654  ENSMUSG00000096808                     2  ENSMUSG00000096808   \n",
       "76129  ENSMUSG00000144196                     1  ENSMUSG00000144196   \n",
       "76130  ENSMUSG00000144199                     1  ENSMUSG00000144199   \n",
       "76131  ENSMUSG00000144201                     1  ENSMUSG00000144201   \n",
       "76132  ENSMUSG00000144203                     1  ENSMUSG00000144203   \n",
       "76133  ENSMUSG00000144204                     1  ENSMUSG00000144204   \n",
       "76204  ENSMUSG00002074899                     1  ENSMUSG00002074899   \n",
       "76275  ENSMUSG00002074970                     1  ENSMUSG00002074970   \n",
       "76705  ENSMUSG00002075400                     1  ENSMUSG00002075400   \n",
       "77034  ENSMUSG00002075729                     1  ENSMUSG00002075729   \n",
       "78195  ENSMUSG00002076890                     1  ENSMUSG00002076890   \n",
       "\n",
       "      gene_symbol_source_db gene_symbol_source_id          gene_biotype  \\\n",
       "17497            EntrezGene                114564        protein_coding   \n",
       "20614                  None                  None        protein_coding   \n",
       "22214                  None                  None        protein_coding   \n",
       "22215                  None                  None        protein_coding   \n",
       "22218                  None                  None        protein_coding   \n",
       "22441                  None                  None        protein_coding   \n",
       "22442                  None                  None        protein_coding   \n",
       "22444                  None                  None        protein_coding   \n",
       "30066                  None                  None        protein_coding   \n",
       "31226                  None                  None  processed_pseudogene   \n",
       "31286                  None                  None        protein_coding   \n",
       "31359                  None                  None        protein_coding   \n",
       "31379                  None                  None        protein_coding   \n",
       "31399                  None                  None        protein_coding   \n",
       "31439                  None                  None        protein_coding   \n",
       "31460                  None                  None        protein_coding   \n",
       "31569                  None                  None        protein_coding   \n",
       "31572                  None                  None        protein_coding   \n",
       "31581                  None                  None        protein_coding   \n",
       "31610                  None                  None        protein_coding   \n",
       "31616                  None                  None        protein_coding   \n",
       "31635                  None                  None        protein_coding   \n",
       "31645                  None                  None        protein_coding   \n",
       "31656                  None                  None        protein_coding   \n",
       "31666                  None                  None        protein_coding   \n",
       "31678                  None                  None        protein_coding   \n",
       "31728                  None                  None        protein_coding   \n",
       "31738                  None                  None        protein_coding   \n",
       "31760                  None                  None        protein_coding   \n",
       "31836                  None                  None        protein_coding   \n",
       "31940                  None                  None        protein_coding   \n",
       "31952                  None                  None        protein_coding   \n",
       "31966                  None                  None        protein_coding   \n",
       "31968                  None                  None        protein_coding   \n",
       "31979                  None                  None        protein_coding   \n",
       "32008                  None                  None        protein_coding   \n",
       "32069                  None                  None        protein_coding   \n",
       "32073                  None                  None        protein_coding   \n",
       "32112                  None                  None        protein_coding   \n",
       "32123                  None                  None        protein_coding   \n",
       "32289                  None                  None        protein_coding   \n",
       "32358                  None                  None        protein_coding   \n",
       "32359                  None                  None        protein_coding   \n",
       "32509            EntrezGene                545762        protein_coding   \n",
       "32572                  None                  None        protein_coding   \n",
       "32606                  None                  None        protein_coding   \n",
       "32626                  None                  None        protein_coding   \n",
       "32639            EntrezGene             100039479        protein_coding   \n",
       "32654                  None                  None        protein_coding   \n",
       "76129                  None                  None                lncRNA   \n",
       "76130                  None                  None                lncRNA   \n",
       "76131                  None                  None                lncRNA   \n",
       "76132                  None                  None                lncRNA   \n",
       "76133                  None                  None                lncRNA   \n",
       "76204                  None                  None                 miRNA   \n",
       "76275                  None                  None                 miRNA   \n",
       "76705                  None                  None                 miRNA   \n",
       "77034                  None                  None                 miRNA   \n",
       "78195                  None                  None                 miRNA   \n",
       "\n",
       "      ensembl_source ensembl_created_date ensembl_modified_date  \\\n",
       "17497        ensembl  2020-04-27 16:28:04   2020-08-25 18:37:28   \n",
       "20614        ensembl  2020-04-27 16:31:13   2020-08-25 18:37:19   \n",
       "22214        ensembl  2020-04-27 16:17:40   2020-08-25 18:37:25   \n",
       "22215        ensembl  2020-04-27 16:11:51   2020-08-25 18:37:27   \n",
       "22218        ensembl  2020-04-27 16:26:11   2020-08-25 18:37:27   \n",
       "22441        ensembl  2020-04-27 16:13:32   2020-08-25 18:37:27   \n",
       "22442        ensembl  2020-04-27 16:09:17   2020-08-25 18:37:26   \n",
       "22444        ensembl  2020-04-27 16:05:55   2020-08-25 18:37:18   \n",
       "30066        ensembl  2020-04-27 17:39:52   2020-08-25 18:37:22   \n",
       "31226        ensembl  2020-04-27 16:15:58   2020-08-25 18:37:25   \n",
       "31286        ensembl  2020-04-27 17:46:01   2020-08-25 18:37:22   \n",
       "31359        ensembl  2020-04-27 15:58:15   2020-08-25 18:37:19   \n",
       "31379        ensembl  2020-04-27 16:32:12   2020-08-25 18:37:19   \n",
       "31399        ensembl  2020-04-27 17:38:10   2020-08-25 18:37:21   \n",
       "31439        ensembl  2020-04-27 17:44:31   2020-08-25 18:37:21   \n",
       "31460        ensembl  2020-04-27 16:18:46   2020-08-25 18:37:26   \n",
       "31569        ensembl  2020-04-27 15:53:54   2020-08-25 18:37:24   \n",
       "31572        ensembl  2020-04-27 16:27:52   2020-08-25 18:37:28   \n",
       "31581        ensembl  2020-04-27 16:28:44   2020-08-25 18:37:18   \n",
       "31610        ensembl  2020-04-27 17:49:25   2020-08-25 18:37:21   \n",
       "31616        ensembl  2020-04-27 16:15:52   2020-08-25 18:37:27   \n",
       "31635        ensembl  2020-04-27 17:30:28   2020-08-25 18:37:21   \n",
       "31645        ensembl  2020-04-27 17:40:26   2020-08-25 18:37:23   \n",
       "31656        ensembl  2020-04-27 16:27:03   2020-08-25 18:37:28   \n",
       "31666        ensembl  2020-04-27 17:44:31   2020-08-25 18:37:22   \n",
       "31678        ensembl  2020-04-27 16:04:14   2020-08-25 18:37:18   \n",
       "31728        ensembl  2020-04-27 16:37:12   2020-08-25 18:37:24   \n",
       "31738        ensembl  2020-04-27 17:39:20   2020-08-25 18:37:20   \n",
       "31760        ensembl  2020-04-27 16:10:08   2020-08-25 18:37:27   \n",
       "31836        ensembl  2020-04-27 16:12:45   2020-08-25 18:37:25   \n",
       "31940        ensembl  2020-04-27 16:27:03   2020-08-25 18:37:28   \n",
       "31952        ensembl  2020-04-27 17:49:39   2020-08-25 18:37:23   \n",
       "31966        ensembl  2020-04-27 16:27:03   2020-08-25 18:37:28   \n",
       "31968        ensembl  2020-04-27 17:36:37   2020-08-25 18:37:29   \n",
       "31979        ensembl  2020-04-27 17:48:06   2020-08-25 18:37:23   \n",
       "32008        ensembl  2020-04-27 16:33:52   2020-08-25 18:37:19   \n",
       "32069        ensembl  2020-04-27 17:06:55   2020-08-25 18:37:29   \n",
       "32073        ensembl  2020-04-27 16:16:46   2020-08-25 18:37:25   \n",
       "32112        ensembl  2020-04-27 17:38:53   2020-08-25 18:37:22   \n",
       "32123        ensembl  2020-04-27 16:14:23   2020-08-25 18:37:25   \n",
       "32289        ensembl  2020-04-27 16:15:05   2020-08-25 18:37:25   \n",
       "32358        ensembl  2020-04-27 15:59:52   2020-08-25 18:37:20   \n",
       "32359        ensembl  2020-04-27 17:38:48   2020-08-25 18:37:23   \n",
       "32509        ensembl  2020-04-27 17:49:11   2020-08-25 18:37:21   \n",
       "32572        ensembl  2020-04-27 17:40:55   2020-08-25 18:37:29   \n",
       "32606        ensembl  2020-04-27 16:35:31   2020-08-25 18:37:19   \n",
       "32626        ensembl  2020-04-27 16:01:38   2020-08-25 18:37:20   \n",
       "32639        ensembl  2020-04-27 17:37:35   2020-08-25 18:37:23   \n",
       "32654        ensembl  2020-04-27 16:28:04   2020-08-25 18:37:28   \n",
       "76129  havana_tagene  2024-05-06 19:10:14   2024-05-06 19:10:14   \n",
       "76130  havana_tagene  2024-05-06 19:10:14   2024-05-06 19:10:14   \n",
       "76131  havana_tagene  2024-05-06 19:10:14   2024-05-06 19:10:14   \n",
       "76132  havana_tagene  2024-05-06 19:10:14   2024-05-06 19:10:14   \n",
       "76133  havana_tagene  2024-05-06 19:10:14   2024-05-06 19:10:14   \n",
       "76204        ensembl  2020-12-10 05:01:07   2020-12-10 05:01:07   \n",
       "76275        ensembl  2020-12-10 05:01:07   2020-12-10 05:01:07   \n",
       "76705        ensembl  2020-12-10 05:01:07   2020-12-10 05:01:07   \n",
       "77034        ensembl  2020-12-10 05:01:07   2020-12-10 05:01:07   \n",
       "78195        ensembl  2020-12-10 05:01:07   2020-12-10 05:01:07   \n",
       "\n",
       "      coord_system_version  ... seq_region_start seq_region_end  \\\n",
       "17497               GRCm39  ...           111572         163011   \n",
       "20614               GRCm39  ...            25923          27230   \n",
       "22214               GRCm39  ...           167446         196478   \n",
       "22215               GRCm39  ...           123792         124928   \n",
       "22218               GRCm39  ...            17591          30203   \n",
       "22441               GRCm39  ...           135395         136519   \n",
       "22442               GRCm39  ...             9124          58882   \n",
       "22444               GRCm39  ...           128555         150452   \n",
       "30066               GRCm39  ...           716296         720517   \n",
       "31226               GRCm39  ...           113868         114758   \n",
       "31286               GRCm39  ...           562161         564894   \n",
       "31359               GRCm39  ...            19740          22248   \n",
       "31379               GRCm39  ...            82993          86051   \n",
       "31399               GRCm39  ...            90482          93220   \n",
       "31439               GRCm39  ...           173050         176134   \n",
       "31460               GRCm39  ...           226405         229375   \n",
       "31569               GRCm39  ...            90528          93623   \n",
       "31572               GRCm39  ...            58743          59114   \n",
       "31581               GRCm39  ...             6366           8939   \n",
       "31610               GRCm39  ...           314841         317584   \n",
       "31616               GRCm39  ...           147792         149707   \n",
       "31635               GRCm39  ...             4975           7356   \n",
       "31645               GRCm39  ...           921942         924675   \n",
       "31656               GRCm39  ...            30394          32306   \n",
       "31666               GRCm39  ...           587624         625594   \n",
       "31678               GRCm39  ...            31967          34932   \n",
       "31728               GRCm39  ...            81607          82689   \n",
       "31738               GRCm39  ...            52190          59690   \n",
       "31760               GRCm39  ...           108390         110303   \n",
       "31836               GRCm39  ...            30483          32396   \n",
       "31940               GRCm39  ...            46295          46666   \n",
       "31952               GRCm39  ...           910289         913083   \n",
       "31966               GRCm39  ...            32239          47294   \n",
       "31968               GRCm39  ...           195181         197903   \n",
       "31979               GRCm39  ...           837364         840451   \n",
       "32008               GRCm39  ...           146926         149659   \n",
       "32069               GRCm39  ...            41338          44417   \n",
       "32073               GRCm39  ...           160465         162404   \n",
       "32112               GRCm39  ...           806210         809281   \n",
       "32123               GRCm39  ...            54469          54840   \n",
       "32289               GRCm39  ...            66955          67326   \n",
       "32358               GRCm39  ...            31101          33667   \n",
       "32359               GRCm39  ...           149261         150163   \n",
       "32509               GRCm39  ...           477246         480331   \n",
       "32572               GRCm39  ...           130715         133448   \n",
       "32606               GRCm39  ...           158543         161339   \n",
       "32626               GRCm39  ...           108318         111392   \n",
       "32639               GRCm39  ...            94186          95088   \n",
       "32654               GRCm39  ...           151235         152734   \n",
       "76129               GRCm39  ...              631         156515   \n",
       "76130               GRCm39  ...            36491         157386   \n",
       "76131               GRCm39  ...            24951          56818   \n",
       "76132               GRCm39  ...           147544         163165   \n",
       "76133               GRCm39  ...            89064         165170   \n",
       "76204               GRCm39  ...            13262          13382   \n",
       "76275               GRCm39  ...            31243          31343   \n",
       "76705               GRCm39  ...            38438          38535   \n",
       "77034               GRCm39  ...            32719          32818   \n",
       "78195               GRCm39  ...            16623          16721   \n",
       "\n",
       "      seq_region_strand  primary_assembly  lrg_gene_id   mhc  \\\n",
       "17497                -1             False          NaN  None   \n",
       "20614                -1             False          NaN  None   \n",
       "22214                 1             False          NaN  None   \n",
       "22215                 1             False          NaN  None   \n",
       "22218                -1             False          NaN  None   \n",
       "22441                -1             False          NaN  None   \n",
       "22442                -1             False          NaN  None   \n",
       "22444                -1             False          NaN  None   \n",
       "30066                 1             False          NaN  None   \n",
       "31226                 1             False          NaN  None   \n",
       "31286                 1             False          NaN  None   \n",
       "31359                -1             False          NaN  None   \n",
       "31379                -1             False          NaN  None   \n",
       "31399                 1             False          NaN  None   \n",
       "31439                 1             False          NaN  None   \n",
       "31460                -1             False          NaN  None   \n",
       "31569                 1             False          NaN  None   \n",
       "31572                 1             False          NaN  None   \n",
       "31581                 1             False          NaN  None   \n",
       "31610                 1             False          NaN  None   \n",
       "31616                 1             False          NaN  None   \n",
       "31635                 1             False          NaN  None   \n",
       "31645                 1             False          NaN  None   \n",
       "31656                -1             False          NaN  None   \n",
       "31666                 1             False          NaN  None   \n",
       "31678                -1             False          NaN  None   \n",
       "31728                 1             False          NaN  None   \n",
       "31738                -1             False          NaN  None   \n",
       "31760                -1             False          NaN  None   \n",
       "31836                -1             False          NaN  None   \n",
       "31940                -1             False          NaN  None   \n",
       "31952                -1             False          NaN  None   \n",
       "31966                 1             False          NaN  None   \n",
       "31968                -1             False          NaN  None   \n",
       "31979                 1             False          NaN  None   \n",
       "32008                -1             False          NaN  None   \n",
       "32069                -1             False          NaN  None   \n",
       "32073                -1             False          NaN  None   \n",
       "32112                 1             False          NaN  None   \n",
       "32123                -1             False          NaN  None   \n",
       "32289                 1             False          NaN  None   \n",
       "32358                 1             False          NaN  None   \n",
       "32359                -1             False          NaN  None   \n",
       "32509                 1             False          NaN  None   \n",
       "32572                -1             False          NaN  None   \n",
       "32606                 1             False          NaN  None   \n",
       "32626                -1             False          NaN  None   \n",
       "32639                -1             False          NaN  None   \n",
       "32654                 1             False          NaN  None   \n",
       "76129                 1             False          NaN  None   \n",
       "76130                 1             False          NaN  None   \n",
       "76131                -1             False          NaN  None   \n",
       "76132                -1             False          NaN  None   \n",
       "76133                -1             False          NaN  None   \n",
       "76204                -1             False          NaN  None   \n",
       "76275                 1             False          NaN  None   \n",
       "76705                 1             False          NaN  None   \n",
       "77034                 1             False          NaN  None   \n",
       "78195                -1             False          NaN  None   \n",
       "\n",
       "                      gene_description       gene_description_source_db  \\\n",
       "17497            component of Sp100-rs  NCBI gene (formerly Entrezgene)   \n",
       "20614                              NaN                              NaN   \n",
       "22214                              NaN                              NaN   \n",
       "22215                              NaN                              NaN   \n",
       "22218                              NaN                              NaN   \n",
       "22441                              NaN                              NaN   \n",
       "22442                              NaN                              NaN   \n",
       "22444                              NaN                              NaN   \n",
       "30066                              NaN                              NaN   \n",
       "31226                              NaN                              NaN   \n",
       "31286                              NaN                              NaN   \n",
       "31359                              NaN                              NaN   \n",
       "31379                              NaN                              NaN   \n",
       "31399                              NaN                              NaN   \n",
       "31439                              NaN                              NaN   \n",
       "31460                              NaN                              NaN   \n",
       "31569                              NaN                              NaN   \n",
       "31572                              NaN                              NaN   \n",
       "31581                              NaN                              NaN   \n",
       "31610                              NaN                              NaN   \n",
       "31616                              NaN                              NaN   \n",
       "31635                              NaN                              NaN   \n",
       "31645                              NaN                              NaN   \n",
       "31656                              NaN                              NaN   \n",
       "31666                              NaN                              NaN   \n",
       "31678                              NaN                              NaN   \n",
       "31728                              NaN                              NaN   \n",
       "31738                              NaN                              NaN   \n",
       "31760                              NaN                              NaN   \n",
       "31836                              NaN                              NaN   \n",
       "31940                              NaN                              NaN   \n",
       "31952                              NaN                              NaN   \n",
       "31966                              NaN                              NaN   \n",
       "31968                              NaN                              NaN   \n",
       "31979                              NaN                              NaN   \n",
       "32008                              NaN                              NaN   \n",
       "32069                              NaN                              NaN   \n",
       "32073                              NaN                              NaN   \n",
       "32112                              NaN                              NaN   \n",
       "32123                              NaN                              NaN   \n",
       "32289                              NaN                              NaN   \n",
       "32358                              NaN                              NaN   \n",
       "32359  vomeronasal 1 receptor Vmn1r187  NCBI gene (formerly Entrezgene)   \n",
       "32509             predicted gene 16367  NCBI gene (formerly Entrezgene)   \n",
       "32572                              NaN                              NaN   \n",
       "32606                              NaN                              NaN   \n",
       "32626                              NaN                              NaN   \n",
       "32639  vomeronasal 1 receptor Vmn1r186  NCBI gene (formerly Entrezgene)   \n",
       "32654                              NaN                              NaN   \n",
       "76129                 novel transcript                              NaN   \n",
       "76130                 novel transcript                              NaN   \n",
       "76131                 novel transcript                              NaN   \n",
       "76132                 novel transcript                              NaN   \n",
       "76133                 novel transcript                              NaN   \n",
       "76204                              NaN                              NaN   \n",
       "76275                              NaN                              NaN   \n",
       "76705                              NaN                              NaN   \n",
       "77034                              NaN                              NaN   \n",
       "78195                              NaN                              NaN   \n",
       "\n",
       "      gene_description_source_id ensembl_representative_gene_id  \n",
       "17497                     114564             ENSMUSG00000062783  \n",
       "20614                        NaN             ENSMUSG00000074720  \n",
       "22214                        NaN             ENSMUSG00000079190  \n",
       "22215                        NaN             ENSMUSG00000079192  \n",
       "22218                        NaN             ENSMUSG00000079222  \n",
       "22441                        NaN             ENSMUSG00000079794  \n",
       "22442                        NaN             ENSMUSG00000079800  \n",
       "22444                        NaN             ENSMUSG00000079808  \n",
       "30066                        NaN             ENSMUSG00000091585  \n",
       "31226                        NaN             ENSMUSG00000094054  \n",
       "31286                        NaN             ENSMUSG00000094172  \n",
       "31359                        NaN             ENSMUSG00000094303  \n",
       "31379                        NaN             ENSMUSG00000094337  \n",
       "31399                        NaN             ENSMUSG00000094383  \n",
       "31439                        NaN             ENSMUSG00000094474  \n",
       "31460                        NaN             ENSMUSG00000094514  \n",
       "31569                        NaN             ENSMUSG00000094722  \n",
       "31572                        NaN             ENSMUSG00000094728  \n",
       "31581                        NaN             ENSMUSG00000094741  \n",
       "31610                        NaN             ENSMUSG00000094791  \n",
       "31616                        NaN             ENSMUSG00000094799  \n",
       "31635                        NaN             ENSMUSG00000094836  \n",
       "31645                        NaN             ENSMUSG00000094855  \n",
       "31656                        NaN             ENSMUSG00000094874  \n",
       "31666                        NaN             ENSMUSG00000094887  \n",
       "31678                        NaN             ENSMUSG00000094915  \n",
       "31728                        NaN             ENSMUSG00000095019  \n",
       "31738                        NaN             ENSMUSG00000095041  \n",
       "31760                        NaN             ENSMUSG00000095092  \n",
       "31836                        NaN             ENSMUSG00000095250  \n",
       "31940                        NaN             ENSMUSG00000095450  \n",
       "31952                        NaN             ENSMUSG00000095475  \n",
       "31966                        NaN             ENSMUSG00000095500  \n",
       "31968                        NaN             ENSMUSG00000095505  \n",
       "31979                        NaN             ENSMUSG00000095523  \n",
       "32008                        NaN             ENSMUSG00000095570  \n",
       "32069                        NaN             ENSMUSG00000095666  \n",
       "32073                        NaN             ENSMUSG00000095672  \n",
       "32112                        NaN             ENSMUSG00000095763  \n",
       "32123                        NaN             ENSMUSG00000095787  \n",
       "32289                        NaN             ENSMUSG00000096100  \n",
       "32358                        NaN             ENSMUSG00000096236  \n",
       "32359                  100039499             ENSMUSG00000096244  \n",
       "32509                     545762             ENSMUSG00000096550  \n",
       "32572                        NaN             ENSMUSG00000096680  \n",
       "32606                        NaN             ENSMUSG00000096728  \n",
       "32626                        NaN             ENSMUSG00000096756  \n",
       "32639                  100039479             ENSMUSG00000096776  \n",
       "32654                        NaN             ENSMUSG00000096808  \n",
       "76129                        NaN             ENSMUSG00000144196  \n",
       "76130                        NaN             ENSMUSG00000144199  \n",
       "76131                        NaN             ENSMUSG00000144201  \n",
       "76132                        NaN             ENSMUSG00000144203  \n",
       "76133                        NaN             ENSMUSG00000144204  \n",
       "76204                        NaN             ENSMUSG00002074899  \n",
       "76275                        NaN             ENSMUSG00002074970  \n",
       "76705                        NaN             ENSMUSG00002075400  \n",
       "77034                        NaN             ENSMUSG00002075729  \n",
       "78195                        NaN             ENSMUSG00002076890  \n",
       "\n",
       "[59 rows x 23 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# many of these genes should probably be alternative alleles rather than representative\n",
    "ensg.gene_df.query(\"not primary_assembly and ensembl_gene_id==ensembl_representative_gene_id\")"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 412.795336,
   "end_time": "2025-01-06T21:16:34.076135",
   "environment_variables": {},
   "exception": null,
   "input_path": "ensembl_genes/notebooks/ensembl_genes_eda.ipynb",
   "output_path": "output/mus_musculus_core_113_39/ensembl_genes_eda.ipynb",
   "parameters": {
    "release": "113",
    "species": "mouse"
   },
   "start_time": "2025-01-06T21:09:41.280799",
   "version": "2.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}